// media-chrome/dist/media-store/state-mediator.js
import { AvailabilityStates, StreamTypes, TextTrackKinds } from "../constants.js";
import { isValidNumber } from "../utils/utils.js";
import { Document, GlobalThis } from "../utils/server-safe-globals.js";
import { containsComposedNode } from "../utils/element-utils.js";
import { getTextTracksList } from "../utils/captions.js";
import { enterFullscreen, exitFullscreen, isFullscreen } from "../utils/fullscreen-api.js";
import { airplaySupported, castSupported, fullscreenSupported, hasFullscreenSupport, hasPipSupport, hasVolumeSupportAsync, pipSupported } from "../utils/platform-tests.js";
import { getShowingSubtitleTracks, getSubtitleTracks, toggleSubtitleTracks } from "./util.js";
//#region node_modules/media-chrome/dist/media-store/state-mediator.js
var StreamTypeValues = Object.values(StreamTypes);
var volumeSupported;
var volumeSupportPromise = hasVolumeSupportAsync().then((supported) => {
	volumeSupported = supported;
	return volumeSupported;
});
var prepareStateOwners = async (...stateOwners) => {
	await Promise.all(stateOwners.filter((x) => x).map(async (stateOwner) => {
		if (!("localName" in stateOwner && stateOwner instanceof GlobalThis.HTMLElement)) return;
		const name = stateOwner.localName;
		if (!name.includes("-")) return;
		const classDef = GlobalThis.customElements.get(name);
		if (classDef && stateOwner instanceof classDef) return;
		await GlobalThis.customElements.whenDefined(name);
		GlobalThis.customElements.upgrade(stateOwner);
	}));
};
var domParser = new GlobalThis.DOMParser();
var parseHtmlToText = (text) => text ? domParser.parseFromString(text, "text/html").body.textContent || text : text;
var stateMediator = {
	mediaError: {
		get(stateOwners, event) {
			const { media } = stateOwners;
			if ((event == null ? void 0 : event.type) === "playing") return;
			return media == null ? void 0 : media.error;
		},
		mediaEvents: [
			"emptied",
			"error",
			"playing"
		]
	},
	mediaErrorCode: {
		get(stateOwners, event) {
			var _a;
			const { media } = stateOwners;
			if ((event == null ? void 0 : event.type) === "playing") return;
			return (_a = media == null ? void 0 : media.error) == null ? void 0 : _a.code;
		},
		mediaEvents: [
			"emptied",
			"error",
			"playing"
		]
	},
	mediaErrorMessage: {
		get(stateOwners, event) {
			var _a, _b;
			const { media } = stateOwners;
			if ((event == null ? void 0 : event.type) === "playing") return;
			return (_b = (_a = media == null ? void 0 : media.error) == null ? void 0 : _a.message) != null ? _b : "";
		},
		mediaEvents: [
			"emptied",
			"error",
			"playing"
		]
	},
	mediaWidth: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.videoWidth) != null ? _a : 0;
		},
		mediaEvents: ["resize"]
	},
	mediaHeight: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.videoHeight) != null ? _a : 0;
		},
		mediaEvents: ["resize"]
	},
	mediaPaused: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.paused) != null ? _a : true;
		},
		set(value, stateOwners) {
			var _a;
			const { media } = stateOwners;
			if (!media) return;
			if (value) media.pause();
			else (_a = media.play()) == null || _a.catch(() => {});
		},
		mediaEvents: [
			"play",
			"playing",
			"pause",
			"emptied"
		]
	},
	mediaHasPlayed: {
		get(stateOwners, event) {
			const { media } = stateOwners;
			if (!media) return false;
			if (!event) return !media.paused;
			return event.type === "playing";
		},
		mediaEvents: ["playing", "emptied"]
	},
	mediaEnded: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.ended) != null ? _a : false;
		},
		mediaEvents: [
			"seeked",
			"ended",
			"emptied"
		]
	},
	mediaPlaybackRate: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.playbackRate) != null ? _a : 1;
		},
		set(value, stateOwners) {
			const { media } = stateOwners;
			if (!media) return;
			if (!Number.isFinite(+value)) return;
			media.playbackRate = +value;
		},
		mediaEvents: ["ratechange", "loadstart"]
	},
	mediaMuted: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.muted) != null ? _a : false;
		},
		set(value, stateOwners) {
			const { media, options: { noMutedPref } = {} } = stateOwners;
			if (!media) return;
			media.muted = value;
			try {
				const hasLocalStoragePrefMuted = GlobalThis.localStorage.getItem("media-chrome-pref-muted") !== null;
				const hasMutedAttribute = media.hasAttribute("muted");
				if (noMutedPref) {
					if (hasLocalStoragePrefMuted) GlobalThis.localStorage.removeItem("media-chrome-pref-muted");
					return;
				}
				if (hasMutedAttribute && !hasLocalStoragePrefMuted) return;
				GlobalThis.localStorage.setItem("media-chrome-pref-muted", value ? "true" : "false");
			} catch (e) {
				console.debug("Error setting muted pref", e);
			}
		},
		mediaEvents: ["volumechange"],
		stateOwnersUpdateHandlers: [(handler, stateOwners) => {
			const { options: { noMutedPref } } = stateOwners;
			const { media } = stateOwners;
			if (!media || media.muted || noMutedPref) return;
			try {
				const mutedPref = GlobalThis.localStorage.getItem("media-chrome-pref-muted") === "true";
				stateMediator.mediaMuted.set(mutedPref, stateOwners);
				handler(mutedPref);
			} catch (e) {
				console.debug("Error getting muted pref", e);
			}
		}]
	},
	mediaLoop: {
		get(stateOwners) {
			const { media } = stateOwners;
			return media == null ? void 0 : media.loop;
		},
		set(value, stateOwners) {
			const { media } = stateOwners;
			if (!media) return;
			media.loop = value;
		},
		mediaEvents: ["medialooprequest"]
	},
	mediaVolume: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.volume) != null ? _a : 1;
		},
		set(value, stateOwners) {
			const { media, options: { noVolumePref } = {} } = stateOwners;
			if (!media) return;
			try {
				if (value == null) GlobalThis.localStorage.removeItem("media-chrome-pref-volume");
				else if (!media.hasAttribute("muted") && !noVolumePref) GlobalThis.localStorage.setItem("media-chrome-pref-volume", value.toString());
			} catch (e) {
				console.debug("Error setting volume pref", e);
			}
			if (!Number.isFinite(+value)) return;
			media.volume = +value;
		},
		mediaEvents: ["volumechange"],
		stateOwnersUpdateHandlers: [(handler, stateOwners) => {
			const { options: { noVolumePref } } = stateOwners;
			if (noVolumePref) return;
			try {
				const { media } = stateOwners;
				if (!media) return;
				const volumePref = GlobalThis.localStorage.getItem("media-chrome-pref-volume");
				if (volumePref == null) return;
				stateMediator.mediaVolume.set(+volumePref, stateOwners);
				handler(+volumePref);
			} catch (e) {
				console.debug("Error getting volume pref", e);
			}
		}]
	},
	mediaVolumeLevel: {
		get(stateOwners) {
			const { media } = stateOwners;
			if (typeof (media == null ? void 0 : media.volume) == "undefined") return "high";
			if (media.muted || media.volume === 0) return "off";
			if (media.volume < .5) return "low";
			if (media.volume < .75) return "medium";
			return "high";
		},
		mediaEvents: ["volumechange"]
	},
	mediaCurrentTime: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return (_a = media == null ? void 0 : media.currentTime) != null ? _a : 0;
		},
		set(value, stateOwners) {
			const { media } = stateOwners;
			if (!media || !isValidNumber(value)) return;
			media.currentTime = value;
		},
		mediaEvents: ["timeupdate", "loadedmetadata"]
	},
	mediaDuration: {
		get(stateOwners) {
			const { media, options: { defaultDuration } = {} } = stateOwners;
			if (defaultDuration && (!media || !media.duration || Number.isNaN(media.duration) || !Number.isFinite(media.duration))) return defaultDuration;
			return Number.isFinite(media == null ? void 0 : media.duration) ? media.duration : NaN;
		},
		mediaEvents: [
			"durationchange",
			"loadedmetadata",
			"emptied"
		]
	},
	mediaLoading: {
		get(stateOwners) {
			const { media } = stateOwners;
			return (media == null ? void 0 : media.readyState) < 3;
		},
		mediaEvents: [
			"waiting",
			"playing",
			"emptied"
		]
	},
	mediaSeekable: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			if (!((_a = media == null ? void 0 : media.seekable) == null ? void 0 : _a.length)) return void 0;
			const start = media.seekable.start(0);
			const end = media.seekable.end(media.seekable.length - 1);
			if (!start && !end) return void 0;
			return [Number(start.toFixed(3)), Number(end.toFixed(3))];
		},
		mediaEvents: [
			"loadedmetadata",
			"emptied",
			"progress",
			"seekablechange"
		]
	},
	mediaBuffered: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			const timeRanges = (_a = media == null ? void 0 : media.buffered) != null ? _a : [];
			return Array.from(timeRanges).map((_, i) => [Number(timeRanges.start(i).toFixed(3)), Number(timeRanges.end(i).toFixed(3))]);
		},
		mediaEvents: ["progress", "emptied"]
	},
	mediaStreamType: {
		get(stateOwners) {
			const { media, options: { defaultStreamType } = {} } = stateOwners;
			const usedDefaultStreamType = [StreamTypes.LIVE, StreamTypes.ON_DEMAND].includes(defaultStreamType) ? defaultStreamType : void 0;
			if (!media) return usedDefaultStreamType;
			const { streamType } = media;
			if (StreamTypeValues.includes(streamType)) {
				if (streamType === StreamTypes.UNKNOWN) return usedDefaultStreamType;
				return streamType;
			}
			const duration = media.duration;
			if (duration === Infinity) return StreamTypes.LIVE;
			else if (Number.isFinite(duration)) return StreamTypes.ON_DEMAND;
			return usedDefaultStreamType;
		},
		mediaEvents: [
			"emptied",
			"durationchange",
			"loadedmetadata",
			"streamtypechange"
		]
	},
	mediaTargetLiveWindow: {
		get(stateOwners) {
			const { media } = stateOwners;
			if (!media) return NaN;
			const { targetLiveWindow } = media;
			const streamType = stateMediator.mediaStreamType.get(stateOwners);
			if ((targetLiveWindow == null || Number.isNaN(targetLiveWindow)) && streamType === StreamTypes.LIVE) return 0;
			return targetLiveWindow;
		},
		mediaEvents: [
			"emptied",
			"durationchange",
			"loadedmetadata",
			"streamtypechange",
			"targetlivewindowchange"
		]
	},
	mediaTimeIsLive: {
		get(stateOwners) {
			const { media, options: { liveEdgeOffset = 10 } = {} } = stateOwners;
			if (!media) return false;
			if (typeof media.liveEdgeStart === "number") {
				if (Number.isNaN(media.liveEdgeStart)) return false;
				return media.currentTime >= media.liveEdgeStart;
			}
			if (!(stateMediator.mediaStreamType.get(stateOwners) === StreamTypes.LIVE)) return false;
			const seekable = media.seekable;
			if (!seekable) return true;
			if (!seekable.length) return false;
			const liveEdgeStart = seekable.end(seekable.length - 1) - liveEdgeOffset;
			return media.currentTime >= liveEdgeStart;
		},
		mediaEvents: [
			"playing",
			"timeupdate",
			"progress",
			"waiting",
			"emptied"
		]
	},
	mediaSubtitlesList: {
		get(stateOwners) {
			return getSubtitleTracks(stateOwners).map(({ kind, label, language }) => ({
				kind,
				label,
				language
			}));
		},
		mediaEvents: ["loadstart"],
		textTracksEvents: ["addtrack", "removetrack"]
	},
	mediaSubtitlesShowing: {
		get(stateOwners) {
			return getShowingSubtitleTracks(stateOwners).map(({ kind, label, language }) => ({
				kind,
				label,
				language
			}));
		},
		mediaEvents: ["loadstart"],
		textTracksEvents: [
			"addtrack",
			"removetrack",
			"change"
		],
		stateOwnersUpdateHandlers: [(_handler, stateOwners) => {
			var _a, _b;
			const { media, options } = stateOwners;
			if (!media) return;
			const updateDefaultSubtitlesCallback = (event) => {
				var _a2;
				if (!options.defaultSubtitles) return;
				if (event && ![TextTrackKinds.CAPTIONS, TextTrackKinds.SUBTITLES].includes((_a2 = event == null ? void 0 : event.track) == null ? void 0 : _a2.kind)) return;
				toggleSubtitleTracks(stateOwners, true);
			};
			media.addEventListener("loadstart", updateDefaultSubtitlesCallback);
			(_a = media.textTracks) == null || _a.addEventListener("addtrack", updateDefaultSubtitlesCallback);
			(_b = media.textTracks) == null || _b.addEventListener("removetrack", updateDefaultSubtitlesCallback);
			return () => {
				var _a2, _b2;
				media.removeEventListener("loadstart", updateDefaultSubtitlesCallback);
				(_a2 = media.textTracks) == null || _a2.removeEventListener("addtrack", updateDefaultSubtitlesCallback);
				(_b2 = media.textTracks) == null || _b2.removeEventListener("removetrack", updateDefaultSubtitlesCallback);
			};
		}]
	},
	mediaChaptersCues: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			if (!media) return [];
			const [chaptersTrack] = getTextTracksList(media, { kind: TextTrackKinds.CHAPTERS });
			return Array.from((_a = chaptersTrack == null ? void 0 : chaptersTrack.cues) != null ? _a : []).map(({ text, startTime, endTime }) => ({
				text: parseHtmlToText(text),
				startTime,
				endTime
			}));
		},
		mediaEvents: ["loadstart", "loadedmetadata"],
		textTracksEvents: [
			"addtrack",
			"removetrack",
			"change"
		],
		stateOwnersUpdateHandlers: [(handler, stateOwners) => {
			var _a;
			const { media } = stateOwners;
			if (!media) return;
			const chaptersTrack = media.querySelector("track[kind=\"chapters\"][default][src]");
			const shadowChaptersTrack = (_a = media.shadowRoot) == null ? void 0 : _a.querySelector(":is(video,audio) > track[kind=\"chapters\"][default][src]");
			chaptersTrack?.addEventListener("load", handler);
			shadowChaptersTrack?.addEventListener("load", handler);
			return () => {
				chaptersTrack?.removeEventListener("load", handler);
				shadowChaptersTrack?.removeEventListener("load", handler);
			};
		}]
	},
	mediaIsPip: {
		get(stateOwners) {
			var _a, _b;
			const { media, documentElement } = stateOwners;
			if (!media || !documentElement) return false;
			if (!documentElement.pictureInPictureElement) return false;
			if (documentElement.pictureInPictureElement === media) return true;
			if (documentElement.pictureInPictureElement instanceof HTMLMediaElement) {
				if (!((_a = media.localName) == null ? void 0 : _a.includes("-"))) return false;
				return containsComposedNode(media, documentElement.pictureInPictureElement);
			}
			if (documentElement.pictureInPictureElement.localName.includes("-")) {
				let currentRoot = documentElement.pictureInPictureElement.shadowRoot;
				while (currentRoot == null ? void 0 : currentRoot.pictureInPictureElement) {
					if (currentRoot.pictureInPictureElement === media) return true;
					currentRoot = (_b = currentRoot.pictureInPictureElement) == null ? void 0 : _b.shadowRoot;
				}
			}
			return false;
		},
		set(value, stateOwners) {
			const { media } = stateOwners;
			if (!media) return;
			if (value) {
				if (!Document.pictureInPictureEnabled) {
					console.warn("MediaChrome: Picture-in-picture is not enabled");
					return;
				}
				if (!media.requestPictureInPicture) {
					console.warn("MediaChrome: The current media does not support picture-in-picture");
					return;
				}
				const warnNotReady = () => {
					console.warn("MediaChrome: The media is not ready for picture-in-picture. It must have a readyState > 0.");
				};
				media.requestPictureInPicture().catch((err) => {
					if (err.code === 11) {
						if (!media.src) {
							console.warn("MediaChrome: The media is not ready for picture-in-picture. It must have a src set.");
							return;
						}
						if (media.readyState === 0 && media.preload === "none") {
							const cleanup = () => {
								media.removeEventListener("loadedmetadata", tryPip);
								media.preload = "none";
							};
							const tryPip = () => {
								media.requestPictureInPicture().catch(warnNotReady);
								cleanup();
							};
							media.addEventListener("loadedmetadata", tryPip);
							media.preload = "metadata";
							setTimeout(() => {
								if (media.readyState === 0) warnNotReady();
								cleanup();
							}, 1e3);
						} else throw err;
					} else throw err;
				});
			} else if (Document.pictureInPictureElement) Document.exitPictureInPicture();
		},
		mediaEvents: ["enterpictureinpicture", "leavepictureinpicture"]
	},
	mediaRenditionList: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return [...(_a = media == null ? void 0 : media.videoRenditions) != null ? _a : []].map((videoRendition) => ({ ...videoRendition }));
		},
		mediaEvents: ["emptied", "loadstart"],
		videoRenditionsEvents: ["addrendition", "removerendition"]
	},
	/** @TODO Model this as a derived value? (CJP) */
	mediaRenditionSelected: {
		get(stateOwners) {
			var _a, _b, _c;
			const { media } = stateOwners;
			return (_c = (_b = media == null ? void 0 : media.videoRenditions) == null ? void 0 : _b[(_a = media.videoRenditions) == null ? void 0 : _a.selectedIndex]) == null ? void 0 : _c.id;
		},
		set(value, stateOwners) {
			const { media } = stateOwners;
			if (!(media == null ? void 0 : media.videoRenditions)) {
				console.warn("MediaController: Rendition selection not supported by this media.");
				return;
			}
			const renditionId = value;
			const index = Array.prototype.findIndex.call(media.videoRenditions, (r) => r.id == renditionId);
			if (media.videoRenditions.selectedIndex != index) media.videoRenditions.selectedIndex = index;
		},
		mediaEvents: ["emptied"],
		videoRenditionsEvents: [
			"addrendition",
			"removerendition",
			"change"
		]
	},
	mediaAudioTrackList: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			return [...(_a = media == null ? void 0 : media.audioTracks) != null ? _a : []];
		},
		mediaEvents: ["emptied", "loadstart"],
		audioTracksEvents: ["addtrack", "removetrack"]
	},
	mediaAudioTrackEnabled: {
		get(stateOwners) {
			var _a, _b;
			const { media } = stateOwners;
			return (_b = [...(_a = media == null ? void 0 : media.audioTracks) != null ? _a : []].find((audioTrack) => audioTrack.enabled)) == null ? void 0 : _b.id;
		},
		set(value, stateOwners) {
			const { media } = stateOwners;
			if (!(media == null ? void 0 : media.audioTracks)) {
				console.warn("MediaChrome: Audio track selection not supported by this media.");
				return;
			}
			const audioTrackId = value;
			for (const track of media.audioTracks) track.enabled = audioTrackId == track.id;
		},
		mediaEvents: ["emptied"],
		audioTracksEvents: [
			"addtrack",
			"removetrack",
			"change"
		]
	},
	mediaIsFullscreen: {
		get(stateOwners) {
			return isFullscreen(stateOwners);
		},
		set(value, stateOwners, event) {
			var _a, _b;
			if (!value) exitFullscreen(stateOwners);
			else {
				enterFullscreen(stateOwners);
				if (event.detail && !((_a = stateOwners.media) == null ? void 0 : _a.inert)) (_b = stateOwners.media) == null || _b.focus();
			}
		},
		rootEvents: ["fullscreenchange", "webkitfullscreenchange"],
		mediaEvents: [
			"webkitbeginfullscreen",
			"webkitendfullscreen",
			"webkitpresentationmodechanged"
		]
	},
	mediaIsCasting: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			if (!(media == null ? void 0 : media.remote) || ((_a = media.remote) == null ? void 0 : _a.state) === "disconnected") return false;
			return !!media.remote.state;
		},
		set(value, stateOwners) {
			var _a, _b;
			const { media } = stateOwners;
			if (!media) return;
			if (value && ((_a = media.remote) == null ? void 0 : _a.state) !== "disconnected") return;
			if (!value && ((_b = media.remote) == null ? void 0 : _b.state) !== "connected") return;
			if (typeof media.remote.prompt !== "function") {
				console.warn("MediaChrome: Casting is not supported in this environment");
				return;
			}
			media.remote.prompt().catch(() => {});
		},
		remoteEvents: [
			"connect",
			"connecting",
			"disconnect"
		]
	},
	mediaIsAirplaying: {
		get() {
			return false;
		},
		set(_value, stateOwners) {
			const { media } = stateOwners;
			if (!media) return;
			if (!(media.webkitShowPlaybackTargetPicker && GlobalThis.WebKitPlaybackTargetAvailabilityEvent)) {
				console.error("MediaChrome: received a request to select AirPlay but AirPlay is not supported in this environment");
				return;
			}
			media.webkitShowPlaybackTargetPicker();
		},
		mediaEvents: ["webkitcurrentplaybacktargetiswirelesschanged"]
	},
	mediaFullscreenUnavailable: { get(stateOwners) {
		const { media } = stateOwners;
		if (!fullscreenSupported || !hasFullscreenSupport(media)) return AvailabilityStates.UNSUPPORTED;
	} },
	mediaPipUnavailable: { get(stateOwners) {
		const { media } = stateOwners;
		if (!pipSupported || !hasPipSupport(media)) return AvailabilityStates.UNSUPPORTED;
		else if (media == null ? void 0 : media.disablePictureInPicture) return AvailabilityStates.UNAVAILABLE;
	} },
	mediaVolumeUnavailable: {
		get(stateOwners) {
			const { media } = stateOwners;
			if (volumeSupported === false || (media == null ? void 0 : media.volume) == void 0) return AvailabilityStates.UNSUPPORTED;
		},
		stateOwnersUpdateHandlers: [(handler) => {
			if (volumeSupported == null) volumeSupportPromise.then((supported) => handler(supported ? void 0 : AvailabilityStates.UNSUPPORTED));
		}]
	},
	mediaCastUnavailable: {
		get(stateOwners, { availability = "not-available" } = {}) {
			var _a;
			const { media } = stateOwners;
			if (!castSupported || !((_a = media == null ? void 0 : media.remote) == null ? void 0 : _a.state)) return AvailabilityStates.UNSUPPORTED;
			if (availability == null || availability === "available") return void 0;
			return AvailabilityStates.UNAVAILABLE;
		},
		stateOwnersUpdateHandlers: [(handler, stateOwners) => {
			var _a;
			const { media } = stateOwners;
			if (!media) return;
			if (!(media.disableRemotePlayback || media.hasAttribute("disableremoteplayback"))) (_a = media == null ? void 0 : media.remote) == null || _a.watchAvailability((availabilityBool) => {
				handler({ availability: availabilityBool ? "available" : "not-available" });
			}).catch((error) => {
				if (error.name === "NotSupportedError") handler({ availability: null });
				else handler({ availability: "not-available" });
			});
			return () => {
				var _a2;
				(_a2 = media == null ? void 0 : media.remote) == null || _a2.cancelWatchAvailability().catch(() => {});
			};
		}]
	},
	mediaAirplayUnavailable: {
		get(_stateOwners, event) {
			if (!airplaySupported) return AvailabilityStates.UNSUPPORTED;
			if ((event == null ? void 0 : event.availability) === "not-available") return AvailabilityStates.UNAVAILABLE;
		},
		mediaEvents: ["webkitplaybacktargetavailabilitychanged"],
		stateOwnersUpdateHandlers: [(handler, stateOwners) => {
			var _a;
			const { media } = stateOwners;
			if (!media) return;
			if (!(media.disableRemotePlayback || media.hasAttribute("disableremoteplayback"))) (_a = media == null ? void 0 : media.remote) == null || _a.watchAvailability((availabilityBool) => {
				handler({ availability: availabilityBool ? "available" : "not-available" });
			}).catch((error) => {
				if (error.name === "NotSupportedError") handler({ availability: null });
				else handler({ availability: "not-available" });
			});
			return () => {
				var _a2;
				(_a2 = media == null ? void 0 : media.remote) == null || _a2.cancelWatchAvailability().catch(() => {});
			};
		}]
	},
	mediaRenditionUnavailable: {
		get(stateOwners) {
			var _a;
			const { media } = stateOwners;
			if (!(media == null ? void 0 : media.videoRenditions)) return AvailabilityStates.UNSUPPORTED;
			if (!((_a = media.videoRenditions) == null ? void 0 : _a.length)) return AvailabilityStates.UNAVAILABLE;
		},
		mediaEvents: ["emptied", "loadstart"],
		videoRenditionsEvents: ["addrendition", "removerendition"]
	},
	mediaAudioTrackUnavailable: {
		get(stateOwners) {
			var _a, _b;
			const { media } = stateOwners;
			if (!(media == null ? void 0 : media.audioTracks)) return AvailabilityStates.UNSUPPORTED;
			if (((_b = (_a = media.audioTracks) == null ? void 0 : _a.length) != null ? _b : 0) <= 1) return AvailabilityStates.UNAVAILABLE;
		},
		mediaEvents: ["emptied", "loadstart"],
		audioTracksEvents: ["addtrack", "removetrack"]
	},
	mediaLang: { get(stateOwners) {
		const { options: { mediaLang } = {} } = stateOwners;
		return mediaLang != null ? mediaLang : "en";
	} }
};
//#endregion
export { prepareStateOwners, stateMediator };
