// @journeyapps/wa-sqlite/src/sqlite-constants.js
var SQLITE_IOERR_ACCESS = 3338;
var SQLITE_IOERR_CHECKRESERVEDLOCK = 3594;
var SQLITE_IOERR_CLOSE = 4106;
var SQLITE_IOERR_DELETE = 2570;
var SQLITE_IOERR_FSTAT = 1802;
var SQLITE_IOERR_FSYNC = 1034;
var SQLITE_IOERR_LOCK = 3850;
var SQLITE_IOERR_TRUNCATE = 1546;
var SQLITE_IOERR_UNLOCK = 2058;
var SQLITE_OPEN_TRANSIENT_DB = 1024;
var SQLITE_OPEN_MAIN_JOURNAL = 2048;
var SQLITE_OPEN_TEMP_JOURNAL = 4096;
var SQLITE_OPEN_SUBJOURNAL = 8192;
var SQLITE_OPEN_SUPER_JOURNAL = 16384;
var SQLITE_OPEN_WAL = 524288;
var SQLITE_IOCAP_UNDELETABLE_WHEN_OPEN = 2048;
var SQLITE_IOCAP_BATCH_ATOMIC = 16384;
//#endregion
export { SQLITE_IOCAP_BATCH_ATOMIC, SQLITE_IOCAP_UNDELETABLE_WHEN_OPEN, SQLITE_IOERR_ACCESS, SQLITE_IOERR_CHECKRESERVEDLOCK, SQLITE_IOERR_CLOSE, SQLITE_IOERR_DELETE, SQLITE_IOERR_FSTAT, SQLITE_IOERR_FSYNC, SQLITE_IOERR_LOCK, SQLITE_IOERR_TRUNCATE, SQLITE_IOERR_UNLOCK, SQLITE_OPEN_MAIN_JOURNAL, SQLITE_OPEN_SUBJOURNAL, SQLITE_OPEN_SUPER_JOURNAL, SQLITE_OPEN_TEMP_JOURNAL, SQLITE_OPEN_TRANSIENT_DB, SQLITE_OPEN_WAL };
