// @journeyapps/wa-sqlite/src/examples/WriteAhead.js
import { Lock } from "./Lock.js";
//#region node_modules/@journeyapps/wa-sqlite/src/examples/WriteAhead.js
var DEFAULT_JOURNAL_SIZE_LIMIT = 1e3;
var DEFAULT_BACKSTOP_INTERVAL = 3e4;
var MAGIC = 931071620;
var FILE_HEADER_SIZE = 32;
var FRAME_HEADER_SIZE = 32;
var FRAME_TYPE_PAGE = 0;
var FRAME_TYPE_COMMIT = 1;
var FRAME_TYPE_END = 2;
/**
* @typedef PageEntry
* @property {number} waOffset location in WAL file
* @property {number} waSalt1 WAL2 file identifier
* @property {number} pageSize
* @property {Uint8Array} [pageData]
*/
/**
* @typedef Transaction
* @property {number} id
* @property {Map<number, PageEntry>} pages address to page data mapping
* @property {number} dbFileSize
* @property {number} [newPageSize]
* @property {number} waSalt1 WAL2 file identifier
* @property {number} waOffsetEnd
*/
/**
* @typedef WriteAheadOptions
* @property {number} [autoCheckpoint]
* @property {number} [backstopInterval]
* @property {number} [journalSizeLimit]
*/
var WriteAhead = class {
	log = null;
	/** @type {WriteAheadOptions} */ options = {
		autoCheckpoint: 1,
		backstopInterval: DEFAULT_BACKSTOP_INTERVAL,
		journalSizeLimit: DEFAULT_JOURNAL_SIZE_LIMIT
	};
	#zName;
	#dbHandle;
	/** @type {FileSystemSyncAccessHandle[]} */ #waHandles;
	/** @type {FileSystemSyncAccessHandle} */ #activeHandle;
	/** @type {{nextTxId: number, salt1: number, salt2: number}} */ #activeHeader;
	/** @type {number} */ #activeOffset;
	/** @type {number} */ #txId = 0;
	/** @type {Transaction} */ #txInProgress = null;
	#dbFileSize = 0;
	/** @type {Promise<any>} */ #ready;
	/** @type {'read'|'write'} */ #isolationState = null;
	/** @type {Lock} */ #txIdLock = null;
	/** @type {Map<number, PageEntry>} */ #waOverlay = /* @__PURE__ */ new Map();
	/** @type {Map<number, Transaction>} */ #mapIdToTx = /* @__PURE__ */ new Map();
	/** @type {Map<number, Transaction>} */ #mapIdToPendingTx = /* @__PURE__ */ new Map();
	#approxPageCount = 0;
	/** @type {BroadcastChannel} */ #broadcastChannel;
	/** @type {number} */ #backstopTimer;
	/** @type {number} */ #backstopTimestamp = 0;
	#abortController = new AbortController();
	/**
	* @param {string} zName
	* @param {FileSystemSyncAccessHandle} dbHandle
	* @param {FileSystemSyncAccessHandle[]} waHandles
	* @param {WriteAheadOptions} options
	*/
	constructor(zName, dbHandle, waHandles, options = {}) {
		this.#zName = zName;
		this.#dbHandle = dbHandle;
		this.#waHandles = waHandles;
		this.options = Object.assign(this.options, options);
		this.#ready = (async () => {
			await this.#updateTxIdLock();
			this.#broadcastChannel = new BroadcastChannel(`${zName}#wa`);
			this.#broadcastChannel.onmessage = (event) => {
				this.#handleMessage(event);
			};
			const fileHeader = this.#waHandles.map((handle) => this.#readFileHeader(handle)).filter((h) => h).sort((a, b) => a.nextTxId - b.nextTxId)[0] ?? this.#writeFileHeader(Math.floor(Math.random() * 4294967295));
			this.#activeHeader = fileHeader;
			this.#activeHandle = this.#waHandles[fileHeader.salt1 & 1];
			this.#activeOffset = FILE_HEADER_SIZE;
			this.#txId = fileHeader.nextTxId - 1;
			for (const tx of this.#readAllTx()) this.#activateTx(tx);
			this.#updateTxIdLock();
			this.#backstopTimestamp = performance.now();
			this.#backstop();
		})();
	}
	/**
	* @returns {Promise<void>}
	*/
	ready() {
		return this.#ready;
	}
	close() {
		this.#abortController.abort();
		this.#broadcastChannel.onmessage = null;
		clearTimeout(this.#backstopTimer);
		this.#txIdLock?.release();
		this.#broadcastChannel.close();
	}
	/**
	* Freeze our view of the database.
	* The view includes the transactions received so far but is not
	* guaranteed to be completely up to date. Unfreeze the view with rejoin().
	*/
	isolateForRead() {
		if (this.#isolationState !== null) throw new Error("Already in isolated state");
		this.#isolationState = "read";
		clearTimeout(this.#backstopTimer);
		this.#backstopTimer = null;
	}
	/**
	* Freeze our view of the database for writing.
	* The view includes all transactions. Unfreeze the view with rejoin().
	*/
	isolateForWrite() {
		if (this.#isolationState !== null) throw new Error("Already in isolated state");
		this.#isolationState = "write";
		clearTimeout(this.#backstopTimer);
		this.#backstopTimer = null;
		this.#advanceTxId({ readToCurrent: true });
	}
	rejoin() {
		if (this.#isolationState === "read") this.#advanceTxId({ autoCheckpoint: true });
		this.#isolationState = null;
		this.#backstop();
	}
	/**
	* @param {number} offset
	* @return {Uint8Array?}
	*/
	read(offset) {
		const pageEntry = this.#txInProgress?.pages.get(offset) ?? this.#waOverlay.get(offset);
		if (pageEntry) {
			if (pageEntry.pageData) {
				this.log?.(`%cread page at ${offset} from WAL ${pageEntry.waSalt1 & 1}:${pageEntry.waOffset} (cached)`, "background-color: gold;");
				return pageEntry.pageData;
			}
			this.log?.(`%cread page at ${offset} from WAL ${pageEntry.waSalt1 & 1}:${pageEntry.waOffset}`, "background-color: gold;");
			return this.#fetchPage(pageEntry);
		}
		return null;
	}
	/**
	* @param {number} offset
	* @param {Uint8Array} data
	* @param {{dstPageSize: number?}} options
	*/
	write(offset, data, options) {
		if (this.#isolationState !== "write") throw new Error("Not in write isolated state");
		if (!this.#txInProgress) {
			const nPageThreshold = this.options.journalSizeLimit > 0 ? this.options.journalSizeLimit : DEFAULT_JOURNAL_SIZE_LIMIT;
			if (this.#approxPageCount >= nPageThreshold && this.#isInactiveFileEmpty()) {
				this.log?.(`%cchange WAL file at ${this.#approxPageCount} pages`, "background-color: lightskyblue;");
				this.#swapActiveFile();
			}
			this.#beginTx();
			if (options.dstPageSize !== data.byteLength) this.#txInProgress.newPageSize = options.dstPageSize;
		}
		if (this.#txInProgress.newPageSize) {
			const frameSize = FRAME_HEADER_SIZE + this.#txInProgress.newPageSize;
			if (data.byteLength > this.#txInProgress.newPageSize) for (let i = 0; i < data.byteLength; i += this.#txInProgress.newPageSize) {
				const pageData = data.slice(i, i + this.#txInProgress.newPageSize);
				const waOffset = this.#writePage(offset + i, pageData);
				this.log?.(`%cwrite page at ${offset + i} to WAL ${this.#activeHeader.salt1 & 1}:${waOffset}`, "background-color: lightskyblue;");
			}
			else {
				const pageOffset = offset % this.#txInProgress.newPageSize;
				const waOffset = this.#activeOffset + (offset - pageOffset) / this.#txInProgress.newPageSize * frameSize + FRAME_HEADER_SIZE + pageOffset;
				this.#activeHandle.write(data.subarray(), { at: waOffset });
				this.log?.(`%cwrite page at ${offset} to WAL ${this.#activeHeader.salt1 & 1}:${waOffset}`, "background-color: lightskyblue;");
			}
		} else {
			const waOffset = this.#writePage(offset, data.slice());
			this.log?.(`%cwrite page at ${offset} to WAL ${this.#activeHeader.salt1 & 1}:${waOffset}`, "background-color: lightskyblue;");
		}
	}
	/**
	* @param {number} newSize
	*/
	truncate(newSize) {
		if (this.#txInProgress) {
			for (const offset of this.#txInProgress.pages.keys()) if (offset >= newSize) this.#txInProgress.pages.delete(offset);
		}
	}
	getFileSize() {
		return this.#txInProgress?.dbFileSize ?? this.#dbFileSize;
	}
	commit() {
		const tx = this.#txInProgress;
		if (tx.newPageSize && tx.pages.size === 0) {
			let pageCount = 1;
			for (let i = 0; i < pageCount; i++) {
				const pageData = new Uint8Array(tx.newPageSize);
				const waOffset = this.#activeOffset + i * (FRAME_HEADER_SIZE + tx.newPageSize) + FRAME_HEADER_SIZE;
				this.#activeHandle.read(pageData, { at: waOffset });
				if (i === 0) pageCount = new DataView(pageData.buffer).getUint32(28);
				this.#writePage(i * tx.newPageSize, pageData);
			}
		}
		const page1 = this.#txInProgress.pages.get(0)?.pageData;
		if (page1) {
			const pageCount = new DataView(page1.buffer, page1.byteOffset, page1.byteLength).getUint32(28);
			this.#txInProgress.dbFileSize = pageCount * page1.byteLength;
		} else {
			this.rollback();
			return;
		}
		this.#commitTx();
		this.#activateTx(tx);
		this.#updateTxIdLock();
		const payload = {
			type: "tx",
			tx
		};
		this.#broadcastChannel.postMessage(payload);
		this.#autoCheckpoint();
		this.#backstopTimestamp = performance.now();
	}
	rollback() {
		this.#abortTx();
	}
	/**
	* @param {{durability: 'strict'|'relaxed'}} options
	*/
	sync(options) {
		if (options.durability === "strict") this.#flushActiveFile();
	}
	/**
	* Move pages from write-ahead to main database file.
	*
	* @param {{isPassive: boolean}} options
	*/
	async checkpoint(options = { isPassive: true }) {
		const lockOptions = { ifAvailable: options.isPassive };
		await navigator.locks.request(`${this.#zName}#ckpt`, lockOptions, async (lock) => {
			if (!lock) return;
			if (this.#abortController.signal.aborted) return;
			let ckptId = this.#getActiveFileStartingTxId() - 1;
			if (options.isPassive) {
				if (!this.#mapIdToTx.has(ckptId)) return;
				if ((await this.#getTxIdLocks()).reduce((min, value) => Math.min(min, value.maxTxId), this.#txId) < ckptId) return;
			} else {
				await this.#waitForTxIdLocks((value) => value.maxTxId >= this.#txId);
				ckptId = this.#txId;
			}
			this.log?.(`%ccheckpoint through txId ${ckptId}`, "background-color: lightgreen;");
			this.#flushInactiveFile();
			if (!options.isPassive) this.#flushActiveFile();
			const writtenOffsets = /* @__PURE__ */ new Set();
			let dbFileSize = this.#dbHandle.getSize();
			for (let tx = this.#mapIdToTx.get(ckptId); tx; tx = this.#mapIdToTx.get(tx.id - 1)) {
				if (tx.id === ckptId && dbFileSize !== tx.dbFileSize) {
					dbFileSize = tx.dbFileSize;
					this.#dbHandle.truncate(dbFileSize);
				}
				for (const [offset, pageEntry] of tx.pages) if (offset < dbFileSize && !writtenOffsets.has(offset)) {
					const pageData = pageEntry.pageData ?? this.#fetchPage(pageEntry);
					if (this.#dbHandle.write(pageData, { at: offset }) !== pageData.byteLength) throw new Error("Checkpoint write failed");
					writtenOffsets.add(offset);
					this.log?.(`%ccheckpoint wrote txId ${tx.id} page at ${offset} to database`, "background-color: lightgreen;");
				}
				if (tx.newPageSize) break;
			}
			this.log?.(`%ccheckpoint flush database file`, "background-color: lightgreen;");
			this.#dbHandle.flush();
			this.#broadcastChannel.postMessage({
				type: "ckpt",
				ckptId
			});
			this.#handleCheckpoint(ckptId);
			this.log?.(`%ccheckpoint waiting for connection updates`, "background-color: lightgreen;");
			await this.#waitForTxIdLocks((value) => value.minTxId > ckptId);
			this.#truncateInactiveFile();
			this.log?.(`%ccheckpoint complete`, "background-color: lightgreen;");
		});
	}
	/**
	* Return the approximate number of write-ahead pages. This is the
	* sum of the number of unique page indices for each transaction,
	* so it can be fewer than the number of pages if any transaction
	* contains multiple frames for the same page.
	* @returns {number}
	*/
	getWriteAheadSize() {
		return this.#approxPageCount;
	}
	isTransactionPending() {
		return !!this.#txInProgress;
	}
	setBackstopInterval(intervalMillis) {
		this.options.backstopInterval = intervalMillis;
		if (intervalMillis > 0 && this.#isolationState) this.#backstop();
	}
	/**
	* Incorporate a transaction into our view of the database.
	* @param {Transaction} tx
	*/
	#activateTx(tx) {
		this.#mapIdToTx.set(tx.id, tx);
		this.#approxPageCount += tx.pages.size;
		for (const [offset, pageEntry] of tx.pages) this.#waOverlay.set(offset, pageEntry);
		this.#dbFileSize = tx.dbFileSize;
	}
	/**
	* Advance the local view of the database. By default, advance to the
	* last broadcast transaction. Optionally, also advance through any
	* additional transactions in the WAL file to be fully current.
	*
	* @param {{readToCurrent?: boolean, autoCheckpoint?: boolean}} options
	*/
	#advanceTxId(options = {}) {
		let didAdvance = false;
		while (this.#mapIdToPendingTx.size) {
			const nextTxId = this.#txId + 1;
			let tx;
			if (this.#mapIdToPendingTx.has(nextTxId)) {
				tx = this.#mapIdToPendingTx.get(nextTxId);
				this.#mapIdToPendingTx.delete(tx.id);
				this.#skipTx(tx);
			} else tx = this.#readTx();
			this.#activateTx(tx);
			didAdvance = true;
		}
		if (options.readToCurrent) for (const tx of this.#readAllTx()) {
			this.#activateTx(tx);
			didAdvance = true;
		}
		if (didAdvance) {
			this.#updateTxIdLock();
			if (options.autoCheckpoint) this.#autoCheckpoint();
		}
		if (options.readToCurrent || didAdvance) this.#backstopTimestamp = performance.now();
	}
	#autoCheckpoint() {
		if (this.options.autoCheckpoint > 0) this.checkpoint({ isPassive: true });
	}
	/**
	* After a checkpoint, remove checkpointed pages from write-ahead.
	* The checkpoint may be been done locally or by another connection.
	* @param {number} ckptId
	*/
	#handleCheckpoint(ckptId) {
		this.log?.(`%capply checkpoint through txId ${ckptId}`, "background-color: lightgreen;");
		for (let tx = this.#mapIdToTx.get(ckptId); tx; tx = this.#mapIdToTx.get(tx.id - 1)) {
			for (const [offset, pageEntry] of tx.pages.entries()) if (this.#waOverlay.get(offset) === pageEntry) {
				this.log?.(`%cremove txId ${tx.id} page at offset ${offset}`, "background-color: lightgreen;");
				this.#waOverlay.delete(offset);
			}
			this.#mapIdToTx.delete(tx.id);
			this.#approxPageCount -= tx.pages.size;
		}
		this.#updateTxIdLock();
	}
	/**
	* @param {MessageEvent} event
	*/
	#handleMessage(event) {
		if (event.data.type === "tx") {
			/** @type {Transaction} */ const tx = event.data.tx;
			if (tx.id > this.#txId) {
				this.#mapIdToPendingTx.set(tx.id, tx);
				if (this.#isolationState === null) this.#advanceTxId({ autoCheckpoint: true });
			}
		} else if (event.data.type === "ckpt") {
			/** @type {number} */ const ckptId = event.data.ckptId;
			this.#handleCheckpoint(ckptId);
		}
	}
	/**
	* Periodic check for recovering from lost transaction broadcasts.
	*/
	#backstop() {
		if (this.options.backstopInterval <= 0) return;
		if (this.#isolationState) throw new Error("Backstop was invoked in an isolated state");
		if (performance.now() >= this.#backstopTimestamp + this.options.backstopInterval) {
			const oldTxId = this.#txId;
			this.#advanceTxId({ readToCurrent: true });
			if (this.#txId > oldTxId) this.log?.(`%cbackstop txId ${oldTxId} -> ${this.#txId}`, "background-color: lightyellow;");
			this.#backstopTimestamp = performance.now();
		}
		const delay = this.#backstopTimestamp + this.options.backstopInterval - performance.now();
		clearTimeout(this.#backstopTimer);
		this.#backstopTimer = self.setTimeout(() => {
			this.#backstop();
		}, delay);
	}
	/**
	* Update the lock that publishes our current txId.
	*/
	async #updateTxIdLock() {
		const oldLock = this.#txIdLock;
		const newLockName = this.#encodeTxIdLockName();
		if (oldLock?.name !== newLockName) {
			this.#txIdLock = new Lock(newLockName);
			await this.#txIdLock.acquire("shared").then(() => {
				oldLock?.release();
			});
			if (this.log) {
				const { minTxId, maxTxId } = this.#decodeTxIdLockName(newLockName);
				this.log?.(`%ctxId to ${minTxId}:${maxTxId}`, "background-color: pink;");
			}
		}
	}
	/**
	* Get all txId locks for this database.
	* @returns {Promise<{name: string, minTxId: number, maxTxId: number, encoded: string}[]>}
	*/
	async #getTxIdLocks() {
		const { held } = await navigator.locks.query();
		return held.map((lock) => this.#decodeTxIdLockName(lock.name)).filter((value) => value !== null);
	}
	/**
	* @returns {string}
	*/
	#encodeTxIdLockName() {
		const maxTxId = this.#txId;
		const minTxId = this.#mapIdToTx.keys().next().value ?? maxTxId + 1;
		return `${this.#zName}-txId<${minTxId.toString(36)}:${maxTxId.toString(36)}>`;
	}
	/**
	* @param {string} lockName
	* @returns {{name: string, minTxId: number, maxTxId: number, encoded: string}?}
	*/
	#decodeTxIdLockName(lockName) {
		const match = lockName.match(/^(.*)-txId<([0-9a-z]+):([0-9a-z]+)>$/);
		if (match?.[1] === this.#zName) return {
			name: match[1],
			minTxId: parseInt(match[2], 36),
			maxTxId: parseInt(match[3], 36),
			encoded: lockName
		};
		return null;
	}
	/**
	* Wait for all txId locks that fail the provided predicate.
	* @param {(lock: {name: string, minTxId: number, maxTxId: number}) => boolean} predicate
	*/
	async #waitForTxIdLocks(predicate) {
		/** @type {string[]} */ let failingLockNames = [];
		do {
			if (failingLockNames.length > 0) await Promise.all(failingLockNames.map((name) => navigator.locks.request(name, async () => {})));
			failingLockNames = (await this.#getTxIdLocks()).filter((value) => !predicate(value)).map((value) => value.encoded);
		} while (failingLockNames.length > 0);
	}
	/**
	* @param {PageEntry} pageEntry
	* @returns {Uint8Array}
	*/
	#fetchPage(pageEntry) {
		const accessHandle = this.#waHandles[pageEntry.waSalt1 & 1];
		const pageData = new Uint8Array(pageEntry.pageSize);
		const nBytesRead = accessHandle.read(pageData, { at: pageEntry.waOffset });
		if (nBytesRead !== pageEntry.pageSize) throw new Error(`Short WAL read: expected ${pageEntry.pageSize} bytes, got ${nBytesRead}`);
		return pageData;
	}
	*#readAllTx() {
		while (true) {
			const tx = this.#readTx();
			if (!tx) break;
			yield tx;
		}
	}
	/**
	* @returns {Transaction?}
	*/
	#readTx() {
		/** @type {Transaction} */ const tx = {
			id: 0,
			pages: /* @__PURE__ */ new Map(),
			dbFileSize: 0,
			waSalt1: 0,
			waOffsetEnd: 0
		};
		let offset = this.#activeOffset;
		while (true) {
			const frame = this.#readFrame(offset);
			if (!frame) return null;
			if (frame.frameType === FRAME_TYPE_PAGE) tx.pages.set(frame.pageOffset, {
				pageSize: frame.pageData.byteLength,
				waOffset: offset + FRAME_HEADER_SIZE,
				waSalt1: this.#activeHeader.salt1
			});
			else if (frame.frameType === FRAME_TYPE_COMMIT) {
				this.#txId += 1;
				this.#activeOffset = offset + frame.byteLength;
				tx.id = this.#txId;
				tx.dbFileSize = frame.dbFileSize;
				tx.waSalt1 = this.#activeHeader.salt1;
				tx.newPageSize = frame.flags & 1 ? tx.pages.get(0).pageSize : null;
				tx.waOffsetEnd = this.#activeOffset;
				return tx;
			} else if (frame.frameType === FRAME_TYPE_END) {
				this.#followFileChange(frame.fileHeader);
				offset = this.#activeOffset;
				continue;
			}
			offset += frame.byteLength;
		}
	}
	/**
	* This method is called when transaction(s) have been received by other
	* means than readTx(), e.g. via BroadcastChannel.
	*
	* @param {Transaction} tx
	*/
	#skipTx(tx) {
		if (tx.waSalt1 !== this.#activeHeader.salt1) {
			if (!this.#followFileChange(null)) throw new Error("invalid WAL file");
		}
		this.#txId = tx.id;
		this.#activeOffset = tx.waOffsetEnd;
	}
	/**
	* @param {{overwrite?: boolean}} options
	* @returns {Transaction}
	*/
	#beginTx(options = {}) {
		this.#txInProgress = {
			id: this.#txId + 1,
			pages: /* @__PURE__ */ new Map(),
			dbFileSize: this.#dbFileSize,
			waSalt1: this.#activeHeader.salt1,
			waOffsetEnd: this.#activeOffset
		};
		return this.#txInProgress;
	}
	/**
	* Write a page frame to the WAL file.
	*
	* @param {number} pageOffset
	* @param {Uint8Array} pageData
	*/
	#writePage(pageOffset, pageData) {
		const headerView = new DataView(new ArrayBuffer(FRAME_HEADER_SIZE));
		headerView.setUint8(0, FRAME_TYPE_PAGE);
		headerView.setUint16(2, pageData.byteLength === 65536 ? 1 : pageData.byteLength);
		headerView.setBigUint64(8, BigInt(pageOffset));
		headerView.setUint32(16, this.#activeHeader.salt1);
		headerView.setUint32(20, this.#activeHeader.salt2);
		const checksum = new Checksum();
		checksum.update(new Uint8Array(headerView.buffer, 0, FRAME_HEADER_SIZE - 8));
		checksum.update(pageData);
		headerView.setUint32(24, checksum.s0);
		headerView.setUint32(28, checksum.s1);
		const bytesWritten = this.#activeHandle.write(headerView, { at: this.#txInProgress.waOffsetEnd }) + this.#activeHandle.write(pageData, { at: this.#txInProgress.waOffsetEnd + FRAME_HEADER_SIZE });
		if (bytesWritten !== headerView.byteLength + pageData.byteLength) throw new Error("write failed");
		const pageEntry = {
			pageSize: pageData.byteLength,
			waOffset: this.#txInProgress.waOffsetEnd + FRAME_HEADER_SIZE,
			waSalt1: this.#activeHeader.salt1,
			pageData: pageOffset === 0 ? pageData : void 0
		};
		this.#txInProgress.pages.set(pageOffset, pageEntry);
		this.#txInProgress.waOffsetEnd += bytesWritten;
		return pageEntry.waOffset;
	}
	/**
	* @returns {Transaction}
	*/
	#commitTx() {
		const headerView = new DataView(new ArrayBuffer(FRAME_HEADER_SIZE));
		headerView.setUint8(0, FRAME_TYPE_COMMIT);
		headerView.setUint8(1, this.#txInProgress.newPageSize ? 1 : 0);
		headerView.setBigUint64(8, BigInt(this.#txInProgress.dbFileSize));
		headerView.setUint32(16, this.#activeHeader.salt1);
		headerView.setUint32(20, this.#activeHeader.salt2);
		const checksum = new Checksum();
		checksum.update(new Uint8Array(headerView.buffer, 0, FRAME_HEADER_SIZE - 8));
		headerView.setUint32(24, checksum.s0);
		headerView.setUint32(28, checksum.s1);
		const bytesWritten = this.#activeHandle.write(headerView, { at: this.#txInProgress.waOffsetEnd });
		if (bytesWritten !== headerView.byteLength) throw new Error("write failed");
		this.#txInProgress.waOffsetEnd += bytesWritten;
		const tx = this.#txInProgress;
		this.#txInProgress = null;
		this.#activeOffset = tx.waOffsetEnd;
		this.#txId = tx.id;
		return tx;
	}
	#abortTx() {
		this.#txInProgress = null;
		this.#activeHandle.truncate(this.#activeOffset);
	}
	/**
	* Switch the active WAL file prior to writing the next transaction.
	*/
	#swapActiveFile() {
		const frameView = new DataView(new ArrayBuffer(FRAME_HEADER_SIZE));
		frameView.setUint8(0, FRAME_TYPE_END);
		frameView.setUint32(16, this.#activeHeader.salt1);
		frameView.setUint32(20, this.#activeHeader.salt2);
		const checksum = new Checksum();
		checksum.update(new Uint8Array(frameView.buffer, 0, FRAME_HEADER_SIZE - 8));
		frameView.setUint32(24, checksum.s0);
		frameView.setUint32(28, checksum.s1);
		if (this.#activeHandle.write(frameView, { at: this.#activeOffset }) !== frameView.byteLength) throw new Error("write failed");
		this.#activeHeader = this.#writeFileHeader();
		this.#activeHandle = this.#getInactiveHandle();
		this.#activeOffset = FILE_HEADER_SIZE;
	}
	#getActiveFileStartingTxId() {
		return this.#activeHeader.nextTxId;
	}
	#flushActiveFile() {
		this.#activeHandle.flush();
	}
	#flushInactiveFile() {
		this.#getInactiveHandle().flush();
	}
	#isInactiveFileEmpty() {
		if (this.#mapIdToTx.has(this.#activeHeader.nextTxId - 1)) return false;
		const inactiveHandle = this.#getInactiveHandle();
		if (inactiveHandle.getSize() < FILE_HEADER_SIZE) return true;
		return this.#readFileHeader(inactiveHandle) === null;
	}
	#truncateInactiveFile() {
		this.#getInactiveHandle().truncate(0);
	}
	/**
	* This method is called after reading an end frame to switch to the
	* other WAL file.
	* @param {{nextTxId: number, salt1: number, salt2: number}?} fileHeader
	*/
	#followFileChange(fileHeader) {
		const accessHandle = this.#getInactiveHandle();
		if (!fileHeader) {
			fileHeader = this.#readFileHeader(accessHandle);
			if (fileHeader?.salt1 !== this.#activeHeader.salt1 + 1 >>> 0) return null;
		}
		this.#activeHandle = accessHandle;
		this.#activeHeader = fileHeader;
		this.#activeOffset = FILE_HEADER_SIZE;
		return fileHeader;
	}
	#getInactiveHandle() {
		return this.#activeHandle !== this.#waHandles[0] ? this.#waHandles[0] : this.#waHandles[1];
	}
	/**
	* @param {FileSystemSyncAccessHandle} accessHandle
	*/
	#readFileHeader(accessHandle) {
		const headerView = new DataView(new ArrayBuffer(FILE_HEADER_SIZE));
		if (accessHandle.read(headerView, { at: 0 }) !== headerView.byteLength) return null;
		if (headerView.getUint32(0) !== MAGIC) return null;
		const checksum = new Checksum();
		checksum.update(new Uint8Array(headerView.buffer, 0, FILE_HEADER_SIZE - 8));
		if (!checksum.matches(headerView.getUint32(24), headerView.getUint32(28))) return null;
		return {
			nextTxId: Number(headerView.getBigUint64(8)),
			salt1: headerView.getUint32(16),
			salt2: headerView.getUint32(20)
		};
	}
	/**
	* @param {number} offset
	*/
	#readFrame(offset) {
		const headerView = new DataView(new ArrayBuffer(FRAME_HEADER_SIZE));
		if (this.#activeHandle.read(headerView, { at: offset }) !== headerView.byteLength) return null;
		const frameSalt1 = headerView.getUint32(16);
		const frameSalt2 = headerView.getUint32(20);
		if (frameSalt1 !== this.#activeHeader.salt1 || frameSalt2 !== this.#activeHeader.salt2) return null;
		const payloadSize = ((size) => size === 1 ? 65536 : size)(headerView.getUint16(2));
		/** @type {Uint8Array} */ let payloadData;
		if (payloadSize) {
			payloadData = new Uint8Array(payloadSize);
			if (this.#activeHandle.read(payloadData, { at: offset + FRAME_HEADER_SIZE }) !== payloadSize) return null;
		}
		const checksum = new Checksum();
		checksum.update(new Uint8Array(headerView.buffer, 0, FRAME_HEADER_SIZE - 8));
		if (payloadData) checksum.update(payloadData);
		if (!checksum.matches(headerView.getUint32(24), headerView.getUint32(28))) return null;
		const frameType = headerView.getUint8(0);
		if (frameType === FRAME_TYPE_PAGE) return {
			frameType,
			byteLength: FRAME_HEADER_SIZE + payloadSize,
			pageOffset: Number(headerView.getBigUint64(8)),
			pageData: payloadData
		};
		else if (frameType === FRAME_TYPE_COMMIT) return {
			frameType,
			byteLength: FRAME_HEADER_SIZE,
			flags: headerView.getUint8(1),
			dbFileSize: Number(headerView.getBigUint64(8))
		};
		else if (frameType === FRAME_TYPE_END) {
			const fileHeader = this.#readFileHeader(this.#getInactiveHandle());
			if (fileHeader?.salt1 !== this.#activeHeader.salt1 + 1 >>> 0) return null;
			return {
				frameType,
				byteLength: FRAME_HEADER_SIZE,
				fileHeader
			};
		}
		throw new Error(`Invalid frame type: ${frameType}`);
	}
	#writeFileHeader(prevSalt1 = this.#activeHeader.salt1) {
		const nextTxId = this.#txId + 1;
		const salt1 = prevSalt1 + 1 >>> 0;
		const salt2 = Math.floor(Math.random() * 4294967295) >>> 0;
		const headerView = new DataView(new ArrayBuffer(FILE_HEADER_SIZE));
		headerView.setUint32(0, MAGIC);
		headerView.setBigUint64(8, BigInt(nextTxId));
		headerView.setUint32(16, salt1);
		headerView.setUint32(20, salt2);
		const checksum = new Checksum();
		checksum.update(new Uint8Array(headerView.buffer, 0, FILE_HEADER_SIZE - 8));
		headerView.setUint32(24, checksum.s0);
		headerView.setUint32(28, checksum.s1);
		if (this.#waHandles[salt1 & 1].write(headerView, { at: 0 }) !== headerView.byteLength) throw new Error("write failed");
		return {
			nextTxId,
			salt1,
			salt2
		};
	}
};
var Checksum = class {
	/** @type {number} */ s0 = 0;
	/** @type {number} */ s1 = 0;
	/**
	* @param {ArrayBuffer|ArrayBufferView} data
	*/
	update(data) {
		if (data.byteLength % 8 !== 0) throw new Error("Data must be a multiple of 8 bytes");
		const words = ArrayBuffer.isView(data) ? new Uint32Array(data.buffer, data.byteOffset, data.byteLength / 4) : new Uint32Array(data);
		for (let i = 0; i < words.length; i += 2) {
			this.s0 = this.s0 + words[i] + this.s1 >>> 0;
			this.s1 = this.s1 + words[i + 1] + this.s0 >>> 0;
		}
	}
	matches(s0, s1) {
		return this.s0 === s0 && this.s1 === s1;
	}
};
//#endregion
export { WriteAhead };
