// media-chrome/dist/utils/server-safe-globals.js
//#region node_modules/media-chrome/dist/utils/server-safe-globals.js
var EventTarget = class {
	addEventListener() {}
	removeEventListener() {}
	dispatchEvent() {
		return true;
	}
};
var Node = class extends EventTarget {};
var Element = class extends Node {
	constructor() {
		super(...arguments);
		this.role = null;
	}
};
var ResizeObserver = class {
	observe() {}
	unobserve() {}
	disconnect() {}
};
var documentShim = {
	createElement: function() {
		return new globalThisShim.HTMLElement();
	},
	createElementNS: function() {
		return new globalThisShim.HTMLElement();
	},
	addEventListener() {},
	removeEventListener() {},
	dispatchEvent(_event) {
		return false;
	}
};
var globalThisShim = {
	ResizeObserver,
	document: documentShim,
	Node,
	Element,
	HTMLElement: class HTMLElement extends Element {
		constructor() {
			super(...arguments);
			this.innerHTML = "";
		}
		get content() {
			return new globalThisShim.DocumentFragment();
		}
	},
	DocumentFragment: class DocumentFragment extends EventTarget {},
	customElements: {
		get: function() {},
		define: function() {},
		whenDefined: function() {}
	},
	localStorage: {
		getItem(_key) {
			return null;
		},
		setItem(_key, _value) {},
		removeItem(_key) {}
	},
	CustomEvent: function CustomEvent() {},
	getComputedStyle: function() {},
	navigator: {
		languages: [],
		get userAgent() {
			return "";
		}
	},
	matchMedia(media) {
		return {
			matches: false,
			media
		};
	},
	DOMParser: class DOMParser {
		parseFromString(string, _contentType) {
			return { body: { textContent: string } };
		}
	}
};
var isServer = "global" in globalThis && (globalThis == null ? void 0 : globalThis.global) === globalThis || typeof window === "undefined" || typeof window.customElements === "undefined";
var isShimmed = Object.keys(globalThisShim).every((key) => key in globalThis);
var GlobalThis = isServer && !isShimmed ? globalThisShim : globalThis;
var Document = isServer && !isShimmed ? documentShim : globalThis.document;
//#endregion
export { Document, GlobalThis };
