// super-media-element/super-media-element.js
//#region node_modules/super-media-element/super-media-element.js
/**
* Super Media Element
* Based on https://github.com/muxinc/custom-video-element - Mux - MIT License
*
* The goal is to create an element that works just like the video element
* but can be extended/sub-classed, because native elements cannot be
* extended today across browsers. Support for extending async loaded video
* like API's. e.g. video players.
*/
var Events = [
	"abort",
	"canplay",
	"canplaythrough",
	"durationchange",
	"emptied",
	"encrypted",
	"ended",
	"error",
	"loadeddata",
	"loadedmetadata",
	"loadstart",
	"pause",
	"play",
	"playing",
	"progress",
	"ratechange",
	"seeked",
	"seeking",
	"stalled",
	"suspend",
	"timeupdate",
	"volumechange",
	"waiting",
	"waitingforkey",
	"resize",
	"enterpictureinpicture",
	"leavepictureinpicture",
	"webkitbeginfullscreen",
	"webkitendfullscreen",
	"webkitpresentationmodechanged"
];
var template = globalThis.document?.createElement("template");
if (template) template.innerHTML = `
    <style>
      :host {
        display: inline-block;
        line-height: 0;
      }

      video,
      audio {
        max-width: 100%;
        max-height: 100%;
        min-width: 100%;
        min-height: 100%;
      }
    </style>
    <slot></slot>
  `;
/**
* @see https://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/
*/
var SuperMediaMixin = (superclass, { tag, is }) => {
	const nativeElTest = globalThis.document?.createElement(tag, { is });
	const nativeElProps = nativeElTest ? getNativeElProps(nativeElTest) : [];
	return class SuperMedia extends superclass {
		static Events = Events;
		static template = template;
		static skipAttributes = [];
		static #isDefined;
		static get observedAttributes() {
			SuperMedia.#define();
			return [
				...nativeElTest?.constructor?.observedAttributes ?? [],
				"autopictureinpicture",
				"disablepictureinpicture",
				"disableremoteplayback",
				"autoplay",
				"controls",
				"controlslist",
				"crossorigin",
				"loop",
				"muted",
				"playsinline",
				"poster",
				"preload",
				"src"
			];
		}
		static #define() {
			if (this.#isDefined) return;
			this.#isDefined = true;
			const propsToAttrs = new Set(this.observedAttributes);
			propsToAttrs.delete("muted");
			for (let prop of nativeElProps) {
				if (prop in this.prototype) continue;
				if (typeof nativeElTest[prop] == "function") this.prototype[prop] = function(...args) {
					this.#init();
					const fn = () => {
						if (this.call) return this.call(prop, ...args);
						return this.nativeEl[prop].apply(this.nativeEl, args);
					};
					if (this.loadComplete && !this.isLoaded) return this.loadComplete.then(fn);
					return fn();
				};
				else {
					let config = { get() {
						this.#init();
						let attr = prop.toLowerCase();
						if (propsToAttrs.has(attr)) {
							const val = this.getAttribute(attr);
							return val === null ? false : val === "" ? true : val;
						}
						return this.get?.(prop) ?? this.nativeEl?.[prop] ?? this.#standinEl[prop];
					} };
					if (prop !== prop.toUpperCase()) config.set = async function(val) {
						this.#init();
						let attr = prop.toLowerCase();
						if (propsToAttrs.has(attr)) {
							if (val === true || val === false || val == null) this.toggleAttribute(attr, Boolean(val));
							else this.setAttribute(attr, val);
							return;
						}
						if (this.loadComplete && !this.isLoaded) await this.loadComplete;
						if (this.set) {
							this.set(prop, val);
							return;
						}
						this.nativeEl[prop] = val;
					};
					Object.defineProperty(this.prototype, prop, config);
				}
			}
		}
		#isInit;
		#loadComplete;
		#hasLoaded = false;
		#isLoaded = false;
		#nativeEl;
		#standinEl;
		constructor() {
			super();
			if (!this.shadowRoot) {
				this.attachShadow({ mode: "open" });
				this.shadowRoot.append(this.constructor.template.content.cloneNode(true));
			}
			if (this.load !== SuperMedia.prototype.load) this.loadComplete = new PublicPromise();
		}
		get loadComplete() {
			return this.#loadComplete;
		}
		set loadComplete(promise) {
			this.#isLoaded = false;
			this.#loadComplete = promise;
			promise?.then(() => {
				this.#isLoaded = true;
			});
		}
		get isLoaded() {
			return this.#isLoaded;
		}
		get nativeEl() {
			return this.#nativeEl ?? this.shadowRoot.querySelector(tag) ?? this.querySelector(tag);
		}
		set nativeEl(val) {
			this.#nativeEl = val;
		}
		get defaultMuted() {
			return this.hasAttribute("muted");
		}
		set defaultMuted(val) {
			this.toggleAttribute("muted", Boolean(val));
		}
		get src() {
			return this.getAttribute("src");
		}
		set src(val) {
			this.setAttribute("src", `${val}`);
		}
		get preload() {
			return this.getAttribute("preload") ?? this.nativeEl?.preload;
		}
		set preload(val) {
			this.setAttribute("preload", `${val}`);
		}
		async #init() {
			if (this.#isInit) return;
			this.#isInit = true;
			this.#initStandinEl();
			this.#initNativeEl();
			for (let prop of nativeElProps) this.#upgradeProperty(prop);
			const childMap = /* @__PURE__ */ new Map();
			const slotEl = this.shadowRoot.querySelector("slot:not([name])");
			slotEl?.addEventListener("slotchange", () => {
				const removeNativeChildren = new Map(childMap);
				slotEl.assignedElements().filter((el) => ["track", "source"].includes(el.localName)).forEach(async (el) => {
					removeNativeChildren.delete(el);
					let clone = childMap.get(el);
					if (!clone) {
						clone = el.cloneNode();
						childMap.set(el, clone);
					}
					if (this.loadComplete && !this.isLoaded) await this.loadComplete;
					this.nativeEl.append?.(clone);
				});
				removeNativeChildren.forEach((el) => el.remove());
			});
			for (let type of this.constructor.Events) this.shadowRoot.addEventListener?.(type, (evt) => {
				if (evt.target !== this.nativeEl) return;
				this.dispatchEvent(new CustomEvent(evt.type, { detail: evt.detail }));
			}, true);
		}
		#upgradeProperty(prop) {
			if (Object.prototype.hasOwnProperty.call(this, prop)) {
				const value = this[prop];
				delete this[prop];
				this[prop] = value;
			}
		}
		#initStandinEl() {
			const dummyEl = document.createElement(tag, { is });
			dummyEl.muted = this.hasAttribute("muted");
			for (let { name, value } of this.attributes) dummyEl.setAttribute(name, value);
			this.#standinEl = {};
			for (let name of getNativeElProps(dummyEl)) this.#standinEl[name] = dummyEl[name];
			dummyEl.removeAttribute("src");
			dummyEl.load();
		}
		async #initNativeEl() {
			if (this.loadComplete && !this.isLoaded) await this.loadComplete;
			if (!this.nativeEl) {
				const nativeEl = document.createElement(tag, { is });
				nativeEl.part = tag;
				this.shadowRoot.append(nativeEl);
			}
			this.nativeEl.muted = this.hasAttribute("muted");
		}
		attributeChangedCallback(attrName, oldValue, newValue) {
			this.#init();
			if (attrName === "src" && this.load !== SuperMedia.prototype.load) this.#loadSrc();
			this.#forwardAttribute(attrName, oldValue, newValue);
		}
		async #loadSrc() {
			if (this.#hasLoaded) this.loadComplete = new PublicPromise();
			this.#hasLoaded = true;
			await Promise.resolve();
			await this.load();
			this.loadComplete?.resolve();
			await this.loadComplete;
		}
		async #forwardAttribute(attrName, oldValue, newValue) {
			if (this.loadComplete && !this.isLoaded) await this.loadComplete;
			if ([
				"id",
				"class",
				...this.constructor.skipAttributes
			].includes(attrName)) return;
			if (newValue === null) this.nativeEl.removeAttribute?.(attrName);
			else this.nativeEl.setAttribute?.(attrName, newValue);
		}
		connectedCallback() {
			this.#init();
		}
	};
};
function getNativeElProps(nativeElTest) {
	let nativeElProps = [];
	for (let proto = Object.getPrototypeOf(nativeElTest); proto && proto !== HTMLElement.prototype; proto = Object.getPrototypeOf(proto)) nativeElProps.push(...Object.getOwnPropertyNames(proto));
	return nativeElProps;
}
/**
* A utility to create Promises with convenient public resolve and reject methods.
* @return {Promise}
*/
var PublicPromise = class extends Promise {
	constructor(executor = () => {}) {
		let res, rej;
		super((resolve, reject) => {
			executor(resolve, reject);
			res = resolve;
			rej = reject;
		});
		this.resolve = res;
		this.reject = rej;
	}
};
var SuperVideoElement = globalThis.document ? SuperMediaMixin(HTMLElement, { tag: "video" }) : class {};
globalThis.document && SuperMediaMixin(HTMLElement, { tag: "audio" });
//#endregion
export { SuperVideoElement };
