// micromark/lib/parse.js
import { combineExtensions } from "../../micromark-util-combine-extensions/index.js";
import { content } from "./initialize/content.js";
import { document } from "./initialize/document.js";
import { flow } from "./initialize/flow.js";
import { string, text } from "./initialize/text.js";
import { constructs_exports } from "./constructs.js";
import { createTokenizer } from "./create-tokenizer.js";
//#region node_modules/micromark/lib/parse.js
/**
* @import {
*   Create,
*   FullNormalizedExtension,
*   InitialConstruct,
*   ParseContext,
*   ParseOptions
* } from 'micromark-util-types'
*/
/**
* @param {ParseOptions | null | undefined} [options]
*   Configuration (optional).
* @returns {ParseContext}
*   Parser.
*/
function parse(options) {
	/** @type {ParseContext} */
	const parser = {
		constructs: combineExtensions([constructs_exports, ...(options || {}).extensions || []]),
		content: create(content),
		defined: [],
		document: create(document),
		flow: create(flow),
		lazy: {},
		string: create(string),
		text: create(text)
	};
	return parser;
	/**
	* @param {InitialConstruct} initial
	*   Construct to start with.
	* @returns {Create}
	*   Create a tokenizer.
	*/
	function create(initial) {
		return creator;
		/** @type {Create} */
		function creator(from) {
			return createTokenizer(parser, initial, from);
		}
	}
}
//#endregion
export { parse };
