// @journeyapps/wa-sqlite/src/examples/OPFSWriteAheadVFS.js
import { SQLITE_IOCAP_BATCH_ATOMIC, SQLITE_IOCAP_UNDELETABLE_WHEN_OPEN, SQLITE_IOERR_ACCESS, SQLITE_IOERR_CLOSE, SQLITE_IOERR_DELETE, SQLITE_IOERR_FSTAT, SQLITE_IOERR_FSYNC, SQLITE_IOERR_LOCK, SQLITE_IOERR_TRUNCATE, SQLITE_IOERR_UNLOCK } from "../sqlite-constants.js";
import "../VFS.js";
import { FacadeVFS } from "../FacadeVFS.js";
import { LazyLock } from "./LazyLock.js";
import { WriteAhead } from "./WriteAhead.js";
//#region node_modules/@journeyapps/wa-sqlite/src/examples/OPFSWriteAheadVFS.js
var LIBRARY_FILES_ROOT = ".wa-sqlite";
var DEFAULT_TEMP_FILES = 6;
var finalizationRegistry = new FinalizationRegistry((f) => f());
/**
* @typedef FileEntry
* @property {string} zName
* @property {number} flags
* @property {FileSystemSyncAccessHandle} [accessHandle]

* Main database file properties:
* @property {*} [retryResult]
* @property {FileSystemSyncAccessHandle[]} [waHandles]
* 
* @property {'reserved'|'exclusive'|null} [writeHint]
* @property {'normal'|'exclusive'} [lockingMode]
* @property {number} [lockState] SQLITE_LOCK_*
* @property {LazyLock} [readLock]
* @property {LazyLock} [writeLock]
* @property {'none'|'read'|'write'|'readwrite'} [useLazyLock]
* @property {number} [timeout]
* @property {0|1|2|3} [synchronous]
* @property {number?} [pageSize]
* @property {boolean} [overwrite]
* 
* @property {WriteAhead} [writeAhead]
*/
/**
* @typedef OPFSWriteAheadOptions
* @property {number} [nTmpFiles]
* @property {number} [autoCheckpoint]
* @property {number} [backstopInterval]
*/
var OPFSWriteAheadVFS = class OPFSWriteAheadVFS extends FacadeVFS {
	lastError = null;
	log = null;
	/** @type {Map<number, FileEntry>} */ mapIdToFile = /* @__PURE__ */ new Map();
	/** @type {Map<string, FileEntry>} */ mapPathToFile = /* @__PURE__ */ new Map();
	/** @type {Map<string, FileSystemSyncAccessHandle>} */ boundTempFiles = /* @__PURE__ */ new Map();
	/** @type {Set<FileSystemSyncAccessHandle>} */ unboundTempFiles = /* @__PURE__ */ new Set();
	/** @type {OPFSWriteAheadOptions} */ options = { nTmpFiles: DEFAULT_TEMP_FILES };
	_ready;
	static async create(name, module, options) {
		const vfs = new OPFSWriteAheadVFS(name, module);
		Object.assign(vfs.options, options);
		await vfs.isReady();
		return vfs;
	}
	constructor(name, module) {
		super(name, module);
		this._ready = (async () => {
			let dirHandle = await navigator.storage.getDirectory();
			dirHandle = await dirHandle.getDirectoryHandle(LIBRARY_FILES_ROOT, { create: true });
			for await (const name of dirHandle.keys()) if (name.startsWith(".session-")) await navigator.locks.request(name, { ifAvailable: true }, async (lock) => {
				if (lock) try {
					await dirHandle.removeEntry(name, { recursive: true });
				} catch (e) {}
			});
			const dirName = `.session-${Math.random().toString(16).slice(2)}`;
			await new Promise((resolve) => {
				navigator.locks.request(dirName, () => {
					resolve();
					return new Promise((release) => {
						finalizationRegistry.register(this, release);
					});
				});
			});
			dirHandle = await dirHandle.getDirectoryHandle(dirName, { create: true });
			for (let i = 0; i < this.options.nTmpFiles; i++) {
				const accessHandle = await (await dirHandle.getFileHandle(i.toString(), { create: true })).createSyncAccessHandle();
				finalizationRegistry.register(this, () => accessHandle.close());
				this.unboundTempFiles.add(accessHandle);
			}
		})();
	}
	isReady() {
		return Promise.all([super.isReady(), this._ready]).then(() => true);
	}
	/**
	* @param {string?} zName 
	* @param {number} fileId 
	* @param {number} flags 
	* @param {DataView} pOutFlags 
	* @returns {number}
	*/
	jOpen(zName, fileId, flags, pOutFlags) {
		try {
			if (zName === null) zName = Math.random().toString(16).slice(2);
			const file = this.mapPathToFile.get(zName) ?? {
				zName,
				flags,
				retryResult: null
			};
			this.mapPathToFile.set(zName, file);
			if (flags & 256) {
				if (file.retryResult === null) {
					this._module.retryOps.push(this.#retryOpen(zName, flags, fileId, pOutFlags));
					return 5;
				} else if (file.retryResult instanceof Error) {
					const e = file.retryResult;
					file.retryResult = null;
					throw e;
				}
				file.accessHandle = file.retryResult.accessHandle;
				file.waHandles = file.retryResult.waHandles;
				file.writeAhead = file.retryResult.writeAhead;
				file.retryResult = null;
				file.lockState = 0;
				file.lockingMode = "normal";
				file.readLock = new LazyLock(`${zName}#read`);
				file.writeLock = new LazyLock(`${zName}#write`);
				file.useLazyLock = "readwrite";
				file.timeout = -1;
				file.synchronous = 1;
				file.writeHint = null;
				file.pageSize = null;
				file.overwrite = false;
			} else if (flags & 540672) throw new Error("WAL and super-journal files are not supported");
			else if (file.accessHandle) {} else {
				if (!(flags & 4)) throw new Error("file not found");
				file.accessHandle = this.#openTemporaryFile(zName);
			}
			this.mapIdToFile.set(fileId, file);
			pOutFlags.setInt32(0, flags, true);
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			this.mapPathToFile.delete(zName);
			return 14;
		}
	}
	/**
	* @param {string} zName 
	* @param {number} syncDir 
	* @returns {number}
	*/
	jDelete(zName, syncDir) {
		try {
			if (this.boundTempFiles.has(zName)) {
				const file = this.mapPathToFile.get(zName);
				this.#deleteTemporaryFile(file);
			} else throw new Error(`unexpected file deletion: ${zName}`);
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_DELETE;
		}
	}
	/**
	* @param {string} zName 
	* @param {number} flags 
	* @param {DataView} pResOut 
	* @returns {number}
	*/
	jAccess(zName, flags, pResOut) {
		try {
			const file = this.mapPathToFile.get(zName);
			pResOut.setInt32(0, file ? 1 : 0, true);
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_ACCESS;
		}
	}
	/**
	* @param {number} fileId 
	* @returns {number}
	*/
	jClose(fileId) {
		try {
			const file = this.mapIdToFile.get(fileId);
			if (file?.flags & 256) {
				file.writeAhead.close();
				file.accessHandle.close();
				file.waHandles.forEach((handle) => handle.close());
				this.mapPathToFile.delete(file?.zName);
				file.readLock.close();
				file.writeLock.close();
			} else if (file?.flags & 8) this.#deleteTemporaryFile(file);
			this.mapIdToFile.delete(fileId);
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_CLOSE;
		}
	}
	/**
	* @param {number} fileId 
	* @param {Uint8Array} pData 
	* @param {number} iOffset
	* @returns {number}
	*/
	jRead(fileId, pData, iOffset) {
		try {
			const file = this.mapIdToFile.get(fileId);
			let bytesRead = null;
			if (file.flags & 256) {
				const pageOffset = iOffset < 100 ? iOffset : 0;
				const page = file.writeAhead.read(iOffset - pageOffset);
				if (page) {
					const readData = page.subarray(pageOffset, pageOffset + pData.byteLength);
					pData.set(readData);
					bytesRead = readData.byteLength;
				}
			}
			if (bytesRead === null) bytesRead = file.accessHandle.read(pData.subarray(), { at: iOffset });
			if (bytesRead < pData.byteLength) {
				pData.fill(0, bytesRead);
				return 522;
			}
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return 266;
		}
	}
	/**
	* @param {number} fileId 
	* @param {Uint8Array} pData 
	* @param {number} iOffset
	* @returns {number}
	*/
	jWrite(fileId, pData, iOffset) {
		try {
			const file = this.mapIdToFile.get(fileId);
			if (file.flags & 256) {
				const isPageResize = file.overwrite && file.pageSize !== pData.byteLength;
				file.writeAhead.write(iOffset, pData, { dstPageSize: isPageResize ? file.pageSize : null });
				return 0;
			}
			file.accessHandle.write(pData.subarray(), { at: iOffset });
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return 778;
		}
	}
	/**
	* @param {number} fileId 
	* @param {number} iSize 
	* @returns {number}
	*/
	jTruncate(fileId, iSize) {
		try {
			const file = this.mapIdToFile.get(fileId);
			if (file.flags & 256) {
				file.writeAhead.truncate(iSize);
				return 0;
			}
			file.accessHandle.truncate(iSize);
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_TRUNCATE;
		}
	}
	/**
	* @param {number} fileId 
	* @param {number} flags 
	* @returns {number}
	*/
	jSync(fileId, flags) {
		try {
			const file = this.mapIdToFile.get(fileId);
			if (file.flags & 256) {
				const durability = file.synchronous > 1 ? "strict" : "relaxed";
				file.writeAhead.sync({ durability });
			}
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_FSYNC;
		}
	}
	/**
	* @param {number} fileId 
	* @param {DataView} pSize64 
	* @returns {number}
	*/
	jFileSize(fileId, pSize64) {
		try {
			const file = this.mapIdToFile.get(fileId);
			let size;
			if (file.flags & 256) size = file.writeAhead.getFileSize() || file.accessHandle.getSize();
			else size = file.accessHandle.getSize();
			pSize64.setBigInt64(0, BigInt(size), true);
			return 0;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_FSTAT;
		}
	}
	/**
	* @param {number} pFile 
	* @param {number} lockType 
	* @returns {number|Promise<number>}
	*/
	jLock(pFile, lockType) {
		try {
			const file = this.mapIdToFile.get(pFile);
			if (file.lockState === 0 && lockType === 1) {
				if (file.retryResult === null) {
					if (file.lockingMode === "exclusive") {
						file.retryResult = {};
						this._module.retryOps.push(this.#retryLockWrite(file));
						return 5;
					}
					if (file.writeHint) if (!file.writeLock.acquireIfHeld("exclusive")) {
						file.retryResult = {};
						this._module.retryOps.push(this.#retryLockWrite(file));
						return 5;
					} else file.writeAhead.isolateForWrite();
					else if (!file.readLock.acquireIfHeld("shared")) {
						file.retryResult = {};
						this._module.retryOps.push(this.#retryLockRead(file));
						return 5;
					} else file.writeAhead.isolateForRead();
				} else if (file.retryResult instanceof Error) {
					const e = file.retryResult;
					file.retryResult = null;
					throw e;
				}
				file.retryResult = null;
			} else if (lockType >= 2 && !file.writeLock.mode) throw new Error("Write transaction cannot use BEGIN DEFERRED");
			file.lockState = lockType;
			return 0;
		} catch (e) {
			if (e.name === "TimeoutError") return 5;
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_LOCK;
		}
	}
	/**
	* @param {number} pFile 
	* @param {number} lockType 
	* @returns {number}
	*/
	jUnlock(pFile, lockType) {
		try {
			const file = this.mapIdToFile.get(pFile);
			if (!file.retryResult && lockType === 0) {
				file.writeAhead.rejoin();
				switch (file.useLazyLock) {
					case "none":
						file.writeLock.release();
						file.readLock.release();
						break;
					case "read":
						file.writeLock.release();
						file.readLock.releaseLazy();
						break;
					case "write":
						file.writeLock.releaseLazy();
						file.readLock.release();
						break;
					case "readwrite":
						file.writeLock.releaseLazy();
						file.readLock.releaseLazy();
						break;
				}
				file.writeHint = null;
			}
			file.lockState = lockType;
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return SQLITE_IOERR_UNLOCK;
		}
	}
	/**
	* @param {number} pFile 
	* @param {DataView} pResOut 
	* @returns {number}
	*/
	jCheckReservedLock(pFile, pResOut) {
		console.assert(false, "unexpected");
		pResOut.setInt32(0, 0, true);
		return 0;
	}
	/**
	* @param {number} pFile
	* @param {number} op
	* @param {DataView} pArg
	* @returns {number}
	*/
	jFileControl(pFile, op, pArg) {
		try {
			const file = this.mapIdToFile.get(pFile);
			switch (op) {
				case 14:
					const key = this._module.UTF8ToString(pArg.getUint32(4, true));
					const valueAddress = pArg.getUint32(8, true);
					const value = valueAddress ? this._module.UTF8ToString(valueAddress) : null;
					this.log?.(`PRAGMA ${key} ${value}`);
					switch (key.toLowerCase()) {
						case "experimental_pragma_20251114":
							switch (value) {
								case "1":
									file.writeHint = "reserved";
									break;
								case "2":
									file.writeHint = "exclusive";
									break;
								default: throw new Error(`unexpected write hint value: ${value}`);
							}
							break;
						case "backstop_interval":
							if (value !== null) {
								const millis = parseInt(value);
								file.writeAhead.setBackstopInterval(millis);
							} else {
								const s = file.writeAhead.options.backstopInterval.toString();
								const ptr = this._module._sqlite3_malloc64(s.length + 1);
								this._module.stringToUTF8(s, ptr, s.length + 1);
								pArg.setUint32(0, ptr, true);
							}
							return 0;
						case "busy_timeout":
							if (value !== null) file.timeout = parseInt(value);
							else {
								const s = file.timeout.toString();
								const ptr = this._module._sqlite3_malloc64(s.length + 1);
								this._module.stringToUTF8(s, ptr, s.length + 1);
								pArg.setUint32(0, ptr, true);
							}
							return 0;
						case "journal_size_limit":
							if (value !== null) {
								const nPages = parseInt(value);
								file.writeAhead.options.journalSizeLimit = nPages;
							}
							break;
						case "locking_mode":
							switch (value?.toLowerCase()) {
								case "normal":
									file.lockingMode = "normal";
									break;
								case "exclusive":
									file.lockingMode = "exclusive";
									break;
							}
							break;
						case "page_size":
							if (value !== null) {
								const n = parseInt(value);
								if (n === 1 || n >= 512 && n <= 32768 && (n & n - 1) === 0) file.pageSize = n === 1 ? 65536 : n;
							}
							break;
						case "synchronous":
							if (value !== null) switch (value.toLowerCase()) {
								case "off":
								case "0":
									file.synchronous = 0;
									break;
								case "normal":
								case "1":
									file.synchronous = 1;
									break;
								case "full":
								case "2":
									file.synchronous = 2;
									break;
								case "extra":
								case "3":
									file.synchronous = 3;
									break;
								default: throw new Error(`unexpected synchronous value: ${value}`);
							}
							break;
						case "vfs_trace":
							if (value !== null) {
								this.log = parseInt(value) !== 0 ? console.debug : null;
								file.writeAhead.log = this.log;
							}
							return 0;
						case "wal_autocheckpoint":
							if (value !== null) file.writeAhead.options.autoCheckpoint = parseInt(value);
							break;
						case "wal_checkpoint":
							const checkpointMode = (value ?? "passive").toLowerCase();
							switch (checkpointMode) {
								case "passive":
									this._module.pendingOps.push(this.#pendingCheckpoint(file, checkpointMode));
									break;
								case "full":
								case "restart":
								case "truncate":
									if (file.writeAhead.isTransactionPending()) throw new Error("invalid while a transaction is in progress");
									this._module.pendingOps.push(this.#pendingCheckpoint(file, checkpointMode));
									break;
								case "noop": break;
								default: throw new Error(`unexpected wal_checkpoint mode: ${value}`);
							}
							{
								const s = file.writeAhead.getWriteAheadSize().toString();
								const ptr = this._module._sqlite3_malloc64(s.length + 1);
								this._module.stringToUTF8(s, ptr, s.length + 1);
								pArg.setUint32(0, ptr, true);
							}
							return 0;
						case "lazy_lock":
							if (value !== null) {
								const useLazyLock = value.toLowerCase();
								switch (useLazyLock) {
									case "read":
									case "write":
									case "readwrite":
									case "none":
										file.useLazyLock = useLazyLock;
										break;
									default: throw new Error(`unexpected value for lazy_lock: ${value}`);
								}
							}
							{
								const s = file.useLazyLock;
								const ptr = this._module._sqlite3_malloc64(s.length + 1);
								this._module.stringToUTF8(s, ptr, s.length + 1);
								pArg.setUint32(0, ptr, true);
							}
							return 0;
					}
					break;
				case 31:
				case 32:
					if (file.flags & 256) return 0;
					break;
				case 33:
					if (file.flags & 256) {
						file.writeAhead.rollback();
						return 0;
					}
					break;
				case 21:
					if (file.flags & 256) file.writeAhead.commit();
					break;
				case 11:
					file.overwrite = true;
					break;
			}
		} catch (e) {
			console.error(e.stack);
			this.lastError = e;
			return 10;
		}
		return 12;
	}
	/**
	* @param {number} pFile
	* @returns {number}
	*/
	jDeviceCharacteristics(pFile) {
		return SQLITE_IOCAP_UNDELETABLE_WHEN_OPEN | SQLITE_IOCAP_BATCH_ATOMIC;
	}
	/**
	* @param {Uint8Array} zBuf 
	* @returns {number}
	*/
	jGetLastError(zBuf) {
		if (this.lastError) {
			console.error(this.lastError);
			const outputArray = zBuf.subarray(0, zBuf.byteLength - 1);
			const { written } = new TextEncoder().encodeInto(this.lastError.message, outputArray);
			zBuf[written] = 0;
		}
		return 0;
	}
	/**
	* @param {string} zName 
	* @returns {FileSystemSyncAccessHandle}
	*/
	#openTemporaryFile(zName) {
		if (this.unboundTempFiles.size === 0) throw new Error("no temporary files available");
		const accessHandle = this.unboundTempFiles.values().next().value;
		this.unboundTempFiles.delete(accessHandle);
		this.boundTempFiles.set(zName, accessHandle);
		return accessHandle;
	}
	/**
	* @param {FileEntry} file 
	*/
	#deleteTemporaryFile(file) {
		file.accessHandle.truncate(0);
		this.mapPathToFile.delete(file.zName);
		this.unboundTempFiles.add(file.accessHandle);
		this.boundTempFiles.delete(file.zName);
	}
	/**
	* @param {string} dbName 
	* @param {number} i 
	* @returns {string}
	*/
	#getWriteAheadNameFromDbName(dbName, i) {
		return `${dbName}-wa${i}`;
	}
	/**
	* Asynchronous PRAGMA operation to checkpoint the write-ahead log.
	* @param {FileEntry} file 
	* @param {'passive'|'full'|'restart'|'truncate'} mode 
	*/
	async #pendingCheckpoint(file, mode) {
		const onFinally = [];
		try {
			if (mode !== "passive" && file.lockState === 0) {
				await file.writeLock.acquire("exclusive");
				onFinally.push(() => file.writeLock.release());
				file.writeAhead.isolateForWrite();
				onFinally.push(() => file.writeAhead.rejoin());
			}
			await file.writeAhead.checkpoint({ isPassive: mode === "passive" });
		} catch (e) {
			if (e.name === "AbortError") e.code = 5;
			throw e;
		} finally {
			while (onFinally.length) onFinally.pop()();
		}
	}
	/**
	* @param {FileEntry} file 
	*/
	async #retryLockRead(file) {
		const onError = [];
		try {
			await file.readLock.acquire("shared", file.timeout);
			onError.push(() => file.readLock.release());
			file.writeAhead.isolateForRead();
			file.retryResult = {};
		} catch (e) {
			while (onError.length) onError.pop()();
			file.retryResult = e;
		}
	}
	/**
	* @param {FileEntry} file 
	*/
	async #retryLockWrite(file) {
		const onError = [];
		try {
			if (file.lockingMode === "exclusive") {
				await file.readLock.acquire("exclusive", file.timeout);
				onError.push(() => file.readLock.release());
			}
			await file.writeLock.acquire("exclusive", file.timeout);
			onError.push(() => file.writeLock.release());
			file.writeAhead.isolateForWrite();
			file.retryResult = {};
		} catch (e) {
			while (onError.length) onError.pop()();
			file.retryResult = e;
		}
	}
	/**
	* Handle asynchronous jOpen() tasks.
	* @param {string} zName 
	* @param {number} flags 
	* @param {number} fileId 
	* @param {DataView} pOutFlags 
	* @returns {Promise<void>}
	*/
	async #retryOpen(zName, flags, fileId, pOutFlags) {
		/** @type {(() => void)[]} */ const onError = [];
		const file = this.mapPathToFile.get(zName);
		try {
			await navigator.locks.request(`${zName}#ckpt`, async (lock) => {
				const directoryNames = zName.split("/").filter((d) => d);
				const dbName = directoryNames.pop();
				let dirHandle = await navigator.storage.getDirectory();
				const create = !!(flags & 4);
				for (const directoryName of directoryNames) dirHandle = await dirHandle.getDirectoryHandle(directoryName, { create });
				const isNewDatabase = create && await (async function() {
					try {
						await dirHandle.getFileHandle(dbName);
						return false;
					} catch (e) {
						if (e.name === "NotFoundError") return true;
						throw e;
					}
				})();
				async function openFile(filename, options) {
					const accessHandle = await (await dirHandle.getFileHandle(filename, options)).createSyncAccessHandle({ mode: "readwrite-unsafe" });
					onError.push(() => {
						accessHandle.close();
						if (isNewDatabase) dirHandle.removeEntry(filename);
					});
					return accessHandle;
				}
				const accessHandle = await openFile(dbName, { create });
				const waHandles = await Promise.all([0, 1].map(async (i) => {
					const waHandle = await openFile(this.#getWriteAheadNameFromDbName(dbName, i), { create: true });
					if (isNewDatabase) waHandle.truncate(0);
					return waHandle;
				}));
				const writeAhead = new WriteAhead(zName, accessHandle, waHandles);
				await writeAhead.ready();
				file.retryResult = {
					accessHandle,
					waHandles,
					writeAhead
				};
			});
		} catch (e) {
			while (onError.length) onError.pop()();
			file.retryResult = e;
		}
	}
};
//#endregion
export { OPFSWriteAheadVFS };
