// @radix-ui/react-context-menu/dist/index.js
"use client";
import { createContextScope } from "../../react-context/dist/index.js";
import { composeEventHandlers } from "../../primitive/dist/index.js";
import { useControllableState } from "../../react-use-controllable-state/dist/index.js";
import { Primitive } from "../../react-primitive/dist/index.js";
import { useCallbackRef } from "../../react-use-callback-ref/dist/index.js";
import { Anchor2, Arrow2, CheckboxItem, Content2 as Content2$1, Group, Item2 as Item2$1, ItemIndicator, Label, Portal, RadioGroup, RadioItem, Root3, Separator, Sub, SubContent, SubTrigger, createMenuScope } from "../../react-menu/dist/index.js";
import * as React$1 from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
//#region node_modules/@radix-ui/react-context-menu/dist/index.mjs
var CONTEXT_MENU_NAME = "ContextMenu";
var [createContextMenuContext, createContextMenuScope] = createContextScope(CONTEXT_MENU_NAME, [createMenuScope]);
var useMenuScope = createMenuScope();
var [ContextMenuProvider, useContextMenuContext] = createContextMenuContext(CONTEXT_MENU_NAME);
var ContextMenu = (props) => {
	const { __scopeContextMenu, children, onOpenChange, dir, modal = true } = props;
	const [open, setOpen] = React$1.useState(false);
	const menuScope = useMenuScope(__scopeContextMenu);
	const handleOpenChangeProp = useCallbackRef(onOpenChange);
	const handleOpenChange = React$1.useCallback((open2) => {
		setOpen(open2);
		handleOpenChangeProp(open2);
	}, [handleOpenChangeProp]);
	return /* @__PURE__ */ jsx(ContextMenuProvider, {
		scope: __scopeContextMenu,
		open,
		onOpenChange: handleOpenChange,
		modal,
		children: /* @__PURE__ */ jsx(Root3, {
			...menuScope,
			dir,
			open,
			onOpenChange: handleOpenChange,
			modal,
			children
		})
	});
};
ContextMenu.displayName = CONTEXT_MENU_NAME;
var TRIGGER_NAME = "ContextMenuTrigger";
var ContextMenuTrigger = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, disabled = false, ...triggerProps } = props;
	const context = useContextMenuContext(TRIGGER_NAME, __scopeContextMenu);
	const menuScope = useMenuScope(__scopeContextMenu);
	const pointRef = React$1.useRef({
		x: 0,
		y: 0
	});
	const virtualRef = React$1.useRef({ getBoundingClientRect: () => DOMRect.fromRect({
		width: 0,
		height: 0,
		...pointRef.current
	}) });
	const longPressTimerRef = React$1.useRef(0);
	const clearLongPress = React$1.useCallback(() => window.clearTimeout(longPressTimerRef.current), []);
	const handleOpen = (event) => {
		pointRef.current = {
			x: event.clientX,
			y: event.clientY
		};
		context.onOpenChange(true);
	};
	React$1.useEffect(() => clearLongPress, [clearLongPress]);
	React$1.useEffect(() => void (disabled && clearLongPress()), [disabled, clearLongPress]);
	return /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx(Anchor2, {
		...menuScope,
		virtualRef
	}), /* @__PURE__ */ jsx(Primitive.span, {
		"data-state": context.open ? "open" : "closed",
		"data-disabled": disabled ? "" : void 0,
		...triggerProps,
		ref: forwardedRef,
		style: {
			WebkitTouchCallout: "none",
			...props.style
		},
		onContextMenu: disabled ? props.onContextMenu : composeEventHandlers(props.onContextMenu, (event) => {
			clearLongPress();
			handleOpen(event);
			event.preventDefault();
		}),
		onPointerDown: disabled ? props.onPointerDown : composeEventHandlers(props.onPointerDown, whenTouchOrPen((event) => {
			clearLongPress();
			longPressTimerRef.current = window.setTimeout(() => handleOpen(event), 700);
		})),
		onPointerMove: disabled ? props.onPointerMove : composeEventHandlers(props.onPointerMove, whenTouchOrPen(clearLongPress)),
		onPointerCancel: disabled ? props.onPointerCancel : composeEventHandlers(props.onPointerCancel, whenTouchOrPen(clearLongPress)),
		onPointerUp: disabled ? props.onPointerUp : composeEventHandlers(props.onPointerUp, whenTouchOrPen(clearLongPress))
	})] });
});
ContextMenuTrigger.displayName = TRIGGER_NAME;
var PORTAL_NAME = "ContextMenuPortal";
var ContextMenuPortal = (props) => {
	const { __scopeContextMenu, ...portalProps } = props;
	return /* @__PURE__ */ jsx(Portal, {
		...useMenuScope(__scopeContextMenu),
		...portalProps
	});
};
ContextMenuPortal.displayName = PORTAL_NAME;
var CONTENT_NAME = "ContextMenuContent";
var ContextMenuContent = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...contentProps } = props;
	const context = useContextMenuContext(CONTENT_NAME, __scopeContextMenu);
	const menuScope = useMenuScope(__scopeContextMenu);
	const hasInteractedOutsideRef = React$1.useRef(false);
	return /* @__PURE__ */ jsx(Content2$1, {
		...menuScope,
		...contentProps,
		ref: forwardedRef,
		side: "right",
		sideOffset: 2,
		align: "start",
		onCloseAutoFocus: (event) => {
			props.onCloseAutoFocus?.(event);
			if (!event.defaultPrevented && hasInteractedOutsideRef.current) event.preventDefault();
			hasInteractedOutsideRef.current = false;
		},
		onInteractOutside: (event) => {
			props.onInteractOutside?.(event);
			if (!event.defaultPrevented && !context.modal) hasInteractedOutsideRef.current = true;
		},
		style: {
			...props.style,
			"--radix-context-menu-content-transform-origin": "var(--radix-popper-transform-origin)",
			"--radix-context-menu-content-available-width": "var(--radix-popper-available-width)",
			"--radix-context-menu-content-available-height": "var(--radix-popper-available-height)",
			"--radix-context-menu-trigger-width": "var(--radix-popper-anchor-width)",
			"--radix-context-menu-trigger-height": "var(--radix-popper-anchor-height)"
		}
	});
});
ContextMenuContent.displayName = CONTENT_NAME;
var GROUP_NAME = "ContextMenuGroup";
var ContextMenuGroup = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...groupProps } = props;
	return /* @__PURE__ */ jsx(Group, {
		...useMenuScope(__scopeContextMenu),
		...groupProps,
		ref: forwardedRef
	});
});
ContextMenuGroup.displayName = GROUP_NAME;
var LABEL_NAME = "ContextMenuLabel";
var ContextMenuLabel = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...labelProps } = props;
	return /* @__PURE__ */ jsx(Label, {
		...useMenuScope(__scopeContextMenu),
		...labelProps,
		ref: forwardedRef
	});
});
ContextMenuLabel.displayName = LABEL_NAME;
var ITEM_NAME = "ContextMenuItem";
var ContextMenuItem = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...itemProps } = props;
	return /* @__PURE__ */ jsx(Item2$1, {
		...useMenuScope(__scopeContextMenu),
		...itemProps,
		ref: forwardedRef
	});
});
ContextMenuItem.displayName = ITEM_NAME;
var CHECKBOX_ITEM_NAME = "ContextMenuCheckboxItem";
var ContextMenuCheckboxItem = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...checkboxItemProps } = props;
	return /* @__PURE__ */ jsx(CheckboxItem, {
		...useMenuScope(__scopeContextMenu),
		...checkboxItemProps,
		ref: forwardedRef
	});
});
ContextMenuCheckboxItem.displayName = CHECKBOX_ITEM_NAME;
var RADIO_GROUP_NAME = "ContextMenuRadioGroup";
var ContextMenuRadioGroup = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...radioGroupProps } = props;
	return /* @__PURE__ */ jsx(RadioGroup, {
		...useMenuScope(__scopeContextMenu),
		...radioGroupProps,
		ref: forwardedRef
	});
});
ContextMenuRadioGroup.displayName = RADIO_GROUP_NAME;
var RADIO_ITEM_NAME = "ContextMenuRadioItem";
var ContextMenuRadioItem = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...radioItemProps } = props;
	return /* @__PURE__ */ jsx(RadioItem, {
		...useMenuScope(__scopeContextMenu),
		...radioItemProps,
		ref: forwardedRef
	});
});
ContextMenuRadioItem.displayName = RADIO_ITEM_NAME;
var INDICATOR_NAME = "ContextMenuItemIndicator";
var ContextMenuItemIndicator = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...itemIndicatorProps } = props;
	return /* @__PURE__ */ jsx(ItemIndicator, {
		...useMenuScope(__scopeContextMenu),
		...itemIndicatorProps,
		ref: forwardedRef
	});
});
ContextMenuItemIndicator.displayName = INDICATOR_NAME;
var SEPARATOR_NAME = "ContextMenuSeparator";
var ContextMenuSeparator = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...separatorProps } = props;
	return /* @__PURE__ */ jsx(Separator, {
		...useMenuScope(__scopeContextMenu),
		...separatorProps,
		ref: forwardedRef
	});
});
ContextMenuSeparator.displayName = SEPARATOR_NAME;
var ARROW_NAME = "ContextMenuArrow";
var ContextMenuArrow = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...arrowProps } = props;
	return /* @__PURE__ */ jsx(Arrow2, {
		...useMenuScope(__scopeContextMenu),
		...arrowProps,
		ref: forwardedRef
	});
});
ContextMenuArrow.displayName = ARROW_NAME;
var SUB_NAME = "ContextMenuSub";
var ContextMenuSub = (props) => {
	const { __scopeContextMenu, children, onOpenChange, open: openProp, defaultOpen } = props;
	const menuScope = useMenuScope(__scopeContextMenu);
	const [open, setOpen] = useControllableState({
		prop: openProp,
		defaultProp: defaultOpen ?? false,
		onChange: onOpenChange,
		caller: SUB_NAME
	});
	return /* @__PURE__ */ jsx(Sub, {
		...menuScope,
		open,
		onOpenChange: setOpen,
		children
	});
};
ContextMenuSub.displayName = SUB_NAME;
var SUB_TRIGGER_NAME = "ContextMenuSubTrigger";
var ContextMenuSubTrigger = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...triggerItemProps } = props;
	return /* @__PURE__ */ jsx(SubTrigger, {
		...useMenuScope(__scopeContextMenu),
		...triggerItemProps,
		ref: forwardedRef
	});
});
ContextMenuSubTrigger.displayName = SUB_TRIGGER_NAME;
var SUB_CONTENT_NAME = "ContextMenuSubContent";
var ContextMenuSubContent = React$1.forwardRef((props, forwardedRef) => {
	const { __scopeContextMenu, ...subContentProps } = props;
	return /* @__PURE__ */ jsx(SubContent, {
		...useMenuScope(__scopeContextMenu),
		...subContentProps,
		ref: forwardedRef,
		style: {
			...props.style,
			"--radix-context-menu-content-transform-origin": "var(--radix-popper-transform-origin)",
			"--radix-context-menu-content-available-width": "var(--radix-popper-available-width)",
			"--radix-context-menu-content-available-height": "var(--radix-popper-available-height)",
			"--radix-context-menu-trigger-width": "var(--radix-popper-anchor-width)",
			"--radix-context-menu-trigger-height": "var(--radix-popper-anchor-height)"
		}
	});
});
ContextMenuSubContent.displayName = SUB_CONTENT_NAME;
function whenTouchOrPen(handler) {
	return (event) => event.pointerType !== "mouse" ? handler(event) : void 0;
}
var Root2 = ContextMenu;
var Trigger = ContextMenuTrigger;
var Portal2 = ContextMenuPortal;
var Content2 = ContextMenuContent;
var Group2 = ContextMenuGroup;
var Label2 = ContextMenuLabel;
var Item2 = ContextMenuItem;
var CheckboxItem2 = ContextMenuCheckboxItem;
var RadioGroup2 = ContextMenuRadioGroup;
var RadioItem2 = ContextMenuRadioItem;
var ItemIndicator2 = ContextMenuItemIndicator;
var Separator2 = ContextMenuSeparator;
var Sub2 = ContextMenuSub;
var SubTrigger2 = ContextMenuSubTrigger;
var SubContent2 = ContextMenuSubContent;
//#endregion
export { CheckboxItem2, Content2, Group2, Item2, ItemIndicator2, Label2, Portal2, RadioGroup2, RadioItem2, Root2, Separator2, Sub2, SubContent2, SubTrigger2, Trigger };
