// micromark/lib/constructs.js
import { __exportAll } from "../../../_virtual/_rolldown/runtime.js";
import { attention } from "../../micromark-core-commonmark/lib/attention.js";
import { autolink } from "../../micromark-core-commonmark/lib/autolink.js";
import { blockQuote } from "../../micromark-core-commonmark/lib/block-quote.js";
import { characterEscape } from "../../micromark-core-commonmark/lib/character-escape.js";
import { characterReference } from "../../micromark-core-commonmark/lib/character-reference.js";
import { codeFenced } from "../../micromark-core-commonmark/lib/code-fenced.js";
import { codeIndented } from "../../micromark-core-commonmark/lib/code-indented.js";
import { codeText } from "../../micromark-core-commonmark/lib/code-text.js";
import { definition } from "../../micromark-core-commonmark/lib/definition.js";
import { hardBreakEscape } from "../../micromark-core-commonmark/lib/hard-break-escape.js";
import { headingAtx } from "../../micromark-core-commonmark/lib/heading-atx.js";
import { htmlFlow } from "../../micromark-core-commonmark/lib/html-flow.js";
import { htmlText } from "../../micromark-core-commonmark/lib/html-text.js";
import { labelEnd } from "../../micromark-core-commonmark/lib/label-end.js";
import { labelStartImage } from "../../micromark-core-commonmark/lib/label-start-image.js";
import { labelStartLink } from "../../micromark-core-commonmark/lib/label-start-link.js";
import { lineEnding } from "../../micromark-core-commonmark/lib/line-ending.js";
import { thematicBreak } from "../../micromark-core-commonmark/lib/thematic-break.js";
import { list } from "../../micromark-core-commonmark/lib/list.js";
import { setextUnderline } from "../../micromark-core-commonmark/lib/setext-underline.js";
import { resolver } from "./initialize/text.js";
//#region node_modules/micromark/lib/constructs.js
/**
* @import {Extension} from 'micromark-util-types'
*/
var constructs_exports = /* @__PURE__ */ __exportAll({
	attentionMarkers: () => attentionMarkers,
	contentInitial: () => contentInitial,
	disable: () => disable,
	document: () => document,
	flow: () => flow,
	flowInitial: () => flowInitial,
	insideSpan: () => insideSpan,
	string: () => string,
	text: () => text
});
/** @satisfies {Extension['document']} */
var document = {
	[42]: list,
	[43]: list,
	[45]: list,
	[48]: list,
	[49]: list,
	[50]: list,
	[51]: list,
	[52]: list,
	[53]: list,
	[54]: list,
	[55]: list,
	[56]: list,
	[57]: list,
	[62]: blockQuote
};
/** @satisfies {Extension['contentInitial']} */
var contentInitial = { [91]: definition };
/** @satisfies {Extension['flowInitial']} */
var flowInitial = {
	[-2]: codeIndented,
	[-1]: codeIndented,
	[32]: codeIndented
};
/** @satisfies {Extension['flow']} */
var flow = {
	[35]: headingAtx,
	[42]: thematicBreak,
	[45]: [setextUnderline, thematicBreak],
	[60]: htmlFlow,
	[61]: setextUnderline,
	[95]: thematicBreak,
	[96]: codeFenced,
	[126]: codeFenced
};
/** @satisfies {Extension['string']} */
var string = {
	[38]: characterReference,
	[92]: characterEscape
};
/** @satisfies {Extension['text']} */
var text = {
	[-5]: lineEnding,
	[-4]: lineEnding,
	[-3]: lineEnding,
	[33]: labelStartImage,
	[38]: characterReference,
	[42]: attention,
	[60]: [autolink, htmlText],
	[91]: labelStartLink,
	[92]: [hardBreakEscape, characterEscape],
	[93]: labelEnd,
	[95]: attention,
	[96]: codeText
};
/** @satisfies {Extension['insideSpan']} */
var insideSpan = { null: [attention, resolver] };
/** @satisfies {Extension['attentionMarkers']} */
var attentionMarkers = { null: [42, 95] };
/** @satisfies {Extension['disable']} */
var disable = { null: [] };
//#endregion
export { constructs_exports };
