// react-player/dist/patterns.js
//#region node_modules/react-player/dist/patterns.js
var AUDIO_EXTENSIONS = /\.(m4a|m4b|mp4a|mpga|mp2|mp2a|mp3|m2a|m3a|wav|weba|aac|oga|spx)($|\?)/i;
var VIDEO_EXTENSIONS = /\.(mp4|og[gv]|webm|mov|m4v)(#t=[,\d+]+)?($|\?)/i;
var HLS_EXTENSIONS = /\.(m3u8)($|\?)/i;
var DASH_EXTENSIONS = /\.(mpd)($|\?)/i;
var MATCH_URL_MUX = /stream\.mux\.com\/(?!\w+\.m3u8)(\w+)/;
var MATCH_URL_YOUTUBE = /(?:youtu\.be\/|youtube(?:-nocookie|education)?\.com\/(?:embed\/|v\/|watch\/|watch\?v=|watch\?.+&v=|shorts\/|live\/))((\w|-){11})|youtube\.com\/playlist\?list=|youtube\.com\/user\//;
var MATCH_URL_VIMEO = /vimeo\.com\/(?!progressive_redirect).+/;
var MATCH_URL_WISTIA = /(?:wistia\.(?:com|net)|wi\.st)\/(?:medias|embed)\/(?:iframe\/)?([^?]+)/;
var MATCH_URL_SPOTIFY = /open\.spotify\.com\/(\w+)\/(\w+)/i;
var MATCH_URL_TWITCH = /(?:www\.|go\.)?twitch\.tv\/([a-zA-Z0-9_]+|(videos?\/|\?video=)\d+)($|\?)/;
var MATCH_URL_TIKTOK = /tiktok\.com\/(?:player\/v1\/|share\/video\/|@[^/]+\/video\/)([0-9]+)/;
var canPlayFile = (url, test) => {
	if (Array.isArray(url)) {
		for (const item of url) {
			if (typeof item === "string" && canPlayFile(item, test)) return true;
			if (canPlayFile(item.src, test)) return true;
		}
		return false;
	}
	return test(url);
};
var canPlay = {
	html: (url) => canPlayFile(url, (u) => AUDIO_EXTENSIONS.test(u) || VIDEO_EXTENSIONS.test(u)),
	hls: (url) => canPlayFile(url, (u) => HLS_EXTENSIONS.test(u)),
	dash: (url) => canPlayFile(url, (u) => DASH_EXTENSIONS.test(u)),
	mux: (url) => MATCH_URL_MUX.test(url),
	youtube: (url) => MATCH_URL_YOUTUBE.test(url),
	vimeo: (url) => MATCH_URL_VIMEO.test(url) && !VIDEO_EXTENSIONS.test(url) && !HLS_EXTENSIONS.test(url),
	wistia: (url) => MATCH_URL_WISTIA.test(url),
	spotify: (url) => MATCH_URL_SPOTIFY.test(url),
	twitch: (url) => MATCH_URL_TWITCH.test(url),
	tiktok: (url) => MATCH_URL_TIKTOK.test(url)
};
//#endregion
export { AUDIO_EXTENSIONS, canPlay };
