// @powersync/react/lib/hooks/watched/watch-utils.js
import { usePowerSync } from "../PowerSyncContext.js";
import React from "react";
//#region node_modules/@powersync/react/lib/hooks/watched/watch-utils.js
var checkQueryChanged = (query, options) => {
	let _compiled;
	try {
		_compiled = query.compile();
	} catch (error) {
		return false;
	}
	const compiled = _compiled;
	const stringifiedParams = query.stringifiedParameters ?? JSON.stringify(compiled.parameters);
	const stringifiedOptions = JSON.stringify(options);
	const previousQueryRef = React.useRef({
		sqlStatement: compiled.sql,
		stringifiedParams,
		stringifiedOptions
	});
	if (previousQueryRef.current.sqlStatement !== compiled.sql || previousQueryRef.current.stringifiedParams != stringifiedParams || previousQueryRef.current.stringifiedOptions != stringifiedOptions) {
		previousQueryRef.current.sqlStatement = compiled.sql;
		previousQueryRef.current.stringifiedParams = stringifiedParams;
		previousQueryRef.current.stringifiedOptions = stringifiedOptions;
		return true;
	}
	return false;
};
var constructCompatibleQuery = (query, parameters = [], options) => {
	const powerSync = usePowerSync();
	const stringifiedParameters = React.useMemo(() => JSON.stringify(parameters), [parameters]);
	const parsedQuery = React.useMemo(() => {
		if (typeof query == "string") return {
			compile: () => ({
				sql: query,
				parameters
			}),
			execute: () => powerSync.getAll(query, parameters),
			stringifiedParameters
		};
		else return {
			compile: () => {
				const compiled = query.compile();
				return {
					sql: compiled.sql,
					parameters: [...compiled.parameters]
				};
			},
			execute: () => query.execute()
		};
	}, [
		query,
		powerSync,
		stringifiedParameters
	]);
	return {
		parsedQuery,
		queryChanged: checkQueryChanged(parsedQuery, options)
	};
};
//#endregion
export { constructCompatibleQuery };
