// media-chrome/dist/utils/platform-tests.js
import { delay } from "./utils.js";
import { Document, GlobalThis } from "./server-safe-globals.js";
import { isFullscreenEnabled } from "./fullscreen-api.js";
//#region node_modules/media-chrome/dist/utils/platform-tests.js
var testMediaEl;
var getTestMediaEl = () => {
	var _a, _b;
	if (testMediaEl) return testMediaEl;
	testMediaEl = (_b = (_a = Document) == null ? void 0 : _a.createElement) == null ? void 0 : _b.call(_a, "video");
	return testMediaEl;
};
var hasVolumeSupportAsync = async (mediaEl = getTestMediaEl()) => {
	if (!mediaEl) return false;
	const prevVolume = mediaEl.volume;
	mediaEl.volume = prevVolume / 2 + .1;
	const abortController = new AbortController();
	const volumeSupported = await Promise.race([dispatchedVolumeChange(mediaEl, abortController.signal), volumeChanged(mediaEl, prevVolume)]);
	abortController.abort();
	return volumeSupported;
};
var dispatchedVolumeChange = (mediaEl, signal) => {
	return new Promise((resolve) => {
		mediaEl.addEventListener("volumechange", () => resolve(true), { signal });
	});
};
var volumeChanged = async (mediaEl, prevVolume) => {
	for (let i = 0; i < 10; i++) {
		if (mediaEl.volume === prevVolume) return false;
		await delay(10);
	}
	return mediaEl.volume !== prevVolume;
};
var isSafari = /.*Version\/.*Safari\/.*/.test(GlobalThis.navigator.userAgent);
var hasPipSupport = (mediaEl = getTestMediaEl()) => {
	if (GlobalThis.matchMedia("(display-mode: standalone)").matches && isSafari) return false;
	return typeof (mediaEl == null ? void 0 : mediaEl.requestPictureInPicture) === "function";
};
var hasFullscreenSupport = (mediaEl = getTestMediaEl()) => {
	return isFullscreenEnabled({
		documentElement: Document,
		media: mediaEl
	});
};
var fullscreenSupported = hasFullscreenSupport();
var pipSupported = hasPipSupport();
var airplaySupported = !!GlobalThis.WebKitPlaybackTargetAvailabilityEvent;
var castSupported = !!GlobalThis.chrome;
//#endregion
export { airplaySupported, castSupported, fullscreenSupported, hasFullscreenSupport, hasPipSupport, hasVolumeSupportAsync, pipSupported };
