// fractional-indexing-jittered/lib/index.js
//#region node_modules/fractional-indexing-jittered/lib/index.js
function indexCharacterSet(options) {
	const dicts = createCharSetDicts(options.chars);
	const limits = integerLimits(dicts, options.firstPositive, options.mostPositive, options.mostNegative);
	const jitterRange = options.jitterRange ?? Math.floor(Math.pow(dicts.length, 3) / 5);
	const paddingRange = paddingDict(jitterRange, dicts.length);
	return {
		chars: options.chars,
		byChar: dicts.byChar,
		byCode: dicts.byCode,
		length: dicts.length,
		first: dicts.byCode[0],
		last: dicts.byCode[dicts.length - 1],
		firstPositive: limits.firstPositive,
		mostPositive: limits.mostPositive,
		firstNegative: limits.firstNegative,
		mostNegative: limits.mostNegative,
		jitterRange,
		paddingDict: paddingRange
	};
}
function createCharSetDicts(charSet) {
	const byCode = {};
	const byChar = {};
	const length = charSet.length;
	for (let i = 0; i < length; i++) {
		const char = charSet[i];
		byCode[i] = char;
		byChar[char] = i;
	}
	return {
		byCode,
		byChar,
		length
	};
}
function integerLimits(dicts, firstPositive, mostPositive, mostNegative) {
	const firstPositiveIndex = firstPositive ? dicts.byChar[firstPositive] : Math.ceil(dicts.length / 2);
	const mostPositiveIndex = mostPositive ? dicts.byChar[mostPositive] : dicts.length - 1;
	const mostNegativeIndex = mostNegative ? dicts.byChar[mostNegative] : 0;
	if (firstPositiveIndex === void 0 || mostPositiveIndex === void 0 || mostNegativeIndex === void 0) throw new Error("invalid charSet");
	if (mostPositiveIndex - firstPositiveIndex < 3) throw new Error("mostPositive must be at least 3 characters away from neutral");
	if (firstPositiveIndex - mostNegativeIndex < 3) throw new Error("mostNegative must be at least 3 characters away from neutral");
	return {
		firstPositive: dicts.byCode[firstPositiveIndex],
		mostPositive: dicts.byCode[mostPositiveIndex],
		firstNegative: dicts.byCode[firstPositiveIndex - 1],
		mostNegative: dicts.byCode[mostNegativeIndex]
	};
}
function paddingDict(jitterRange, charSetLength) {
	const paddingDict2 = {};
	for (let i = 0; i < 100; i++) {
		paddingDict2[i] = Math.pow(charSetLength, i);
		if (paddingDict2[i] > jitterRange) break;
	}
	return paddingDict2;
}
var _base62CharSet = null;
function base62CharSet() {
	if (_base62CharSet) return _base62CharSet;
	return _base62CharSet = indexCharacterSet({
		chars: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz",
		firstPositive: "a",
		mostPositive: "z",
		mostNegative: "A"
	});
}
function distanceBetween(a, b, charSet) {
	const indexA = charSet.byChar[a];
	const indexB = charSet.byChar[b];
	return Math.abs(indexA - indexB);
}
function integerLength(head, charSet) {
	const firstChar = head[0];
	if (firstChar > charSet.mostPositive || firstChar < charSet.mostNegative) throw new Error("invalid firstChar on key");
	if (firstChar === charSet.mostPositive) return distanceBetween(firstChar, charSet.firstPositive, charSet) + 1 + integerLengthFromSecondLevel(head.slice(1), "positive", charSet);
	if (firstChar === charSet.mostNegative) return distanceBetween(firstChar, charSet.firstNegative, charSet) + 1 + integerLengthFromSecondLevel(head.slice(1), "negative", charSet);
	if (firstChar >= charSet.firstPositive) return distanceBetween(firstChar, charSet.firstPositive, charSet) + 2;
	else return distanceBetween(firstChar, charSet.firstNegative, charSet) + 2;
}
function integerLengthFromSecondLevel(key, direction, charSet) {
	const firstChar = key[0];
	if (firstChar > charSet.mostPositive || firstChar < charSet.mostNegative) throw new Error("invalid firstChar on key");
	if (firstChar === charSet.mostPositive && direction === "positive") return distanceBetween(firstChar, charSet.mostNegative, charSet) + 1 + integerLengthFromSecondLevel(key.slice(1), direction, charSet);
	if (firstChar === charSet.mostNegative && direction === "negative") return distanceBetween(firstChar, charSet.mostPositive, charSet) + 1 + integerLengthFromSecondLevel(key.slice(1), direction, charSet);
	if (direction === "positive") return distanceBetween(firstChar, charSet.mostNegative, charSet) + 2;
	else return distanceBetween(firstChar, charSet.mostPositive, charSet) + 2;
}
function makeSameLength(a, b, pad, fillChar, forceLength) {
	const max = forceLength ?? Math.max(a.length, b.length);
	if (pad === "start") return [a.padStart(max, fillChar), b.padStart(max, fillChar)];
	return [a.padEnd(max, fillChar), b.padEnd(max, fillChar)];
}
function midPoint(lower, upper, charSet) {
	let [paddedLower, paddedUpper] = makeSameLength(lower, upper, "end", charSet.first);
	let distance = lexicalDistance(paddedLower, paddedUpper, charSet);
	if (distance === 1) {
		paddedLower = paddedLower.padEnd(paddedLower.length + 1, charSet.first);
		distance = charSet.length;
	}
	const mid = encodeToCharSet(Math.floor(distance / 2), charSet);
	return addCharSetKeys(paddedLower, mid, charSet);
}
function lexicalDistance(a, b, charSet) {
	const [lower, upper] = makeSameLength(a, b, "end", charSet.first).sort();
	return decodeCharSetToNumber(subtractCharSetKeys(upper, lower, charSet), charSet);
}
function addCharSetKeys(a, b, charSet) {
	const base = charSet.length;
	const [paddedA, paddedB] = makeSameLength(a, b, "start", charSet.first);
	const result = [];
	let carry = 0;
	for (let i = paddedA.length - 1; i >= 0; i--) {
		const sum = charSet.byChar[paddedA[i]] + charSet.byChar[paddedB[i]] + carry;
		carry = Math.floor(sum / base);
		const remainder = sum % base;
		result.unshift(charSet.byCode[remainder]);
	}
	if (carry > 0) result.unshift(charSet.byCode[carry]);
	return result.join("");
}
function subtractCharSetKeys(a, b, charSet, stripLeadingZeros = true) {
	const base = charSet.length;
	const [paddedA, paddedB] = makeSameLength(a, b, "start", charSet.first);
	const result = [];
	let borrow = 0;
	for (let i = paddedA.length - 1; i >= 0; i--) {
		let digitA = charSet.byChar[paddedA[i]];
		const digitB = charSet.byChar[paddedB[i]] + borrow;
		if (digitA < digitB) {
			borrow = 1;
			digitA += base;
		} else borrow = 0;
		const difference = digitA - digitB;
		result.unshift(charSet.byCode[difference]);
	}
	if (borrow > 0) throw new Error("Subtraction result is negative. Ensure a is greater than or equal to b.");
	while (stripLeadingZeros && result.length > 1 && result[0] === charSet.first) result.shift();
	return result.join("");
}
function incrementKey(key, charSet) {
	return addCharSetKeys(key, charSet.byCode[1], charSet);
}
function decrementKey(key, charSet) {
	return subtractCharSetKeys(key, charSet.byCode[1], charSet, false);
}
function encodeToCharSet(int, charSet) {
	if (int === 0) return charSet.byCode[0];
	let res = "";
	const max = charSet.length;
	while (int > 0) {
		res = charSet.byCode[int % max] + res;
		int = Math.floor(int / max);
	}
	return res;
}
function decodeCharSetToNumber(key, charSet) {
	let res = 0;
	const length = key.length;
	const max = charSet.length;
	for (let i = 0; i < length; i++) res += charSet.byChar[key[i]] * Math.pow(max, length - i - 1);
	return res;
}
function startKey(charSet) {
	return charSet.firstPositive + charSet.byCode[0];
}
function validInteger(integer, charSet) {
	return integerLength(integer, charSet) === integer.length;
}
function validateOrderKey(orderKey, charSet) {
	getIntegerPart(orderKey, charSet);
}
function getIntegerPart(orderKey, charSet) {
	const integerPartLength = integerLength(integerHead(orderKey, charSet), charSet);
	if (integerPartLength > orderKey.length) throw new Error("invalid order key length: " + orderKey);
	return orderKey.slice(0, integerPartLength);
}
function validateInteger(integer, charSet) {
	if (!validInteger(integer, charSet)) throw new Error("invalid integer length: " + integer);
}
function incrementInteger(integer, charSet) {
	validateInteger(integer, charSet);
	const [head, digs] = splitInteger(integer, charSet);
	if (digs.split("").some((d) => d !== charSet.byCode[charSet.length - 1])) return head + incrementKey(digs, charSet);
	return startOnNewHead(incrementIntegerHead(head, charSet), "lower", charSet);
}
function decrementInteger(integer, charSet) {
	validateInteger(integer, charSet);
	const [head, digs] = splitInteger(integer, charSet);
	if (digs.split("").some((d) => d !== charSet.byCode[0])) return head + decrementKey(digs, charSet);
	return startOnNewHead(decrementIntegerHead(head, charSet), "upper", charSet);
}
function integerHead(integer, charSet) {
	let i = 0;
	if (integer[0] === charSet.mostPositive) while (integer[i] === charSet.mostPositive) i = i + 1;
	if (integer[0] === charSet.mostNegative) while (integer[i] === charSet.mostNegative) i = i + 1;
	return integer.slice(0, i + 1);
}
function splitInteger(integer, charSet) {
	const head = integerHead(integer, charSet);
	return [head, integer.slice(head.length)];
}
function incrementIntegerHead(head, charSet) {
	const inPositiveRange = head >= charSet.firstPositive;
	const nextHead = incrementKey(head, charSet);
	const headIsLimitMax = head[head.length - 1] === charSet.mostPositive;
	const nextHeadIsLimitMax = nextHead[nextHead.length - 1] === charSet.mostPositive;
	if (inPositiveRange && nextHeadIsLimitMax) return nextHead + charSet.mostNegative;
	if (!inPositiveRange && headIsLimitMax) return head.slice(0, head.length - 1);
	return nextHead;
}
function decrementIntegerHead(head, charSet) {
	const inPositiveRange = head >= charSet.firstPositive;
	const headIsLimitMin = head[head.length - 1] === charSet.mostNegative;
	if (inPositiveRange && headIsLimitMin) return decrementKey(head.slice(0, head.length - 1), charSet);
	if (!inPositiveRange && headIsLimitMin) return head + charSet.mostPositive;
	return decrementKey(head, charSet);
}
function startOnNewHead(head, limit, charSet) {
	const newLength = integerLength(head, charSet);
	return head + (limit === "upper" ? charSet.byCode[charSet.length - 1] : charSet.byCode[0]).repeat(newLength - head.length);
}
function jitterString(orderKey, charSet) {
	return addCharSetKeys(orderKey, encodeToCharSet(Math.floor(Math.random() * charSet.jitterRange), charSet), charSet);
}
function padAndJitterString(orderKey, numberOfChars, charSet) {
	return jitterString(orderKey.padEnd(orderKey.length + numberOfChars, charSet.first), charSet);
}
function paddingNeededForJitter(orderKey, b, charSet) {
	const nextInteger = incrementInteger(getIntegerPart(orderKey, charSet), charSet);
	let needed = 0;
	if (b !== null) {
		const distanceToB = lexicalDistance(orderKey, b, charSet);
		if (distanceToB < charSet.jitterRange + 1) needed = Math.max(needed, paddingNeededForDistance(distanceToB, charSet));
	}
	const distanceToNextInteger = lexicalDistance(orderKey, nextInteger, charSet);
	if (distanceToNextInteger < charSet.jitterRange + 1) needed = Math.max(needed, paddingNeededForDistance(distanceToNextInteger, charSet));
	return needed;
}
function paddingNeededForDistance(distance, charSet) {
	const gap = charSet.jitterRange - distance;
	const firstBigger = Object.entries(charSet.paddingDict).find(([_key, value]) => {
		return value > gap;
	});
	return firstBigger ? parseInt(firstBigger[0]) : 0;
}
function generateKeyBetween(lower, upper, charSet = base62CharSet()) {
	if (lower !== null) validateOrderKey(lower, charSet);
	if (upper !== null) validateOrderKey(upper, charSet);
	if (lower === null && upper === null) return startKey(charSet);
	if (lower === null) return decrementInteger(getIntegerPart(upper, charSet), charSet);
	if (upper === null) return incrementInteger(getIntegerPart(lower, charSet), charSet);
	if (lower >= upper) throw new Error(lower + " >= " + upper);
	return midPoint(lower, upper, charSet);
}
function generateJitteredKeyBetween(lower, upper, charSet = base62CharSet()) {
	const key = generateKeyBetween(lower, upper, charSet);
	const paddingNeeded = paddingNeededForJitter(key, upper, charSet);
	if (paddingNeeded) return padAndJitterString(key, paddingNeeded, charSet);
	return jitterString(key, charSet);
}
function generateNJitteredKeysBetween(lower, upper, n, charSet = base62CharSet()) {
	return spreadGeneratorResults(lower, upper, n, charSet, generateJitteredKeyBetween, generateNJitteredKeysBetween);
}
function spreadGeneratorResults(lower, upper, n, charSet, generateKey, generateNKeys) {
	if (n === 0) return [];
	if (n === 1) return [generateKey(lower, upper, charSet)];
	if (upper == null) {
		let newUpper = generateKey(lower, upper, charSet);
		const result = [newUpper];
		for (let i = 0; i < n - 1; i++) {
			newUpper = generateKey(newUpper, upper, charSet);
			result.push(newUpper);
		}
		return result;
	}
	if (lower == null) {
		let newLower = generateKey(lower, upper, charSet);
		const result = [newLower];
		for (let i = 0; i < n - 1; i++) {
			newLower = generateKey(lower, newLower, charSet);
			result.push(newLower);
		}
		result.reverse();
		return result;
	}
	const mid = Math.floor(n / 2);
	const midOrderKey = generateKey(lower, upper, charSet);
	return [
		...generateNKeys(lower, midOrderKey, mid, charSet),
		midOrderKey,
		...generateNKeys(midOrderKey, upper, n - mid - 1, charSet)
	];
}
//#endregion
export { generateJitteredKeyBetween, generateNJitteredKeysBetween };
