// media-chrome/dist/menu/media-rendition-menu.js
import { MediaUIAttributes, MediaUIEvents } from "../constants.js";
import { parseRenditionList } from "../utils/utils.js";
import { t } from "../utils/i18n.js";
import { GlobalThis } from "../utils/server-safe-globals.js";
import { getMediaController, getNumericAttr, getStringAttr, setNumericAttr, setStringAttr } from "../utils/element-utils.js";
import { MediaChromeMenu, createIndicator, createMenuItem } from "./media-chrome-menu.js";
//#region node_modules/media-chrome/dist/menu/media-rendition-menu.js
var __accessCheck = (obj, member, msg) => {
	if (!member.has(obj)) throw TypeError("Cannot " + msg);
};
var __privateGet = (obj, member, getter) => {
	__accessCheck(obj, member, "read from private field");
	return getter ? getter.call(obj) : member.get(obj);
};
var __privateAdd = (obj, member, value) => {
	if (member.has(obj)) throw TypeError("Cannot add the same private member more than once");
	member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
};
var __privateSet = (obj, member, value, setter) => {
	__accessCheck(obj, member, "write to private field");
	setter ? setter.call(obj, value) : member.set(obj, value);
	return value;
};
var __privateMethod = (obj, member, method) => {
	__accessCheck(obj, member, "access private method");
	return method;
};
var _renditionList, _prevState, _render, render_fn, _onChange, onChange_fn;
var MediaRenditionMenu = class extends MediaChromeMenu {
	constructor() {
		super(...arguments);
		__privateAdd(this, _render);
		__privateAdd(this, _onChange);
		__privateAdd(this, _renditionList, []);
		__privateAdd(this, _prevState, {});
	}
	static get observedAttributes() {
		return [
			...super.observedAttributes,
			MediaUIAttributes.MEDIA_RENDITION_LIST,
			MediaUIAttributes.MEDIA_RENDITION_SELECTED,
			MediaUIAttributes.MEDIA_RENDITION_UNAVAILABLE,
			MediaUIAttributes.MEDIA_HEIGHT,
			MediaUIAttributes.MEDIA_WIDTH
		];
	}
	static formatMenuItemText(text, rendition) {
		return super.formatMenuItemText(text, rendition);
	}
	static formatRendition(rendition, { showBitrate = false } = {}) {
		const renditionText = `${Math.min(rendition.width, rendition.height)}p`;
		if (showBitrate && rendition.bitrate) {
			const mbps = rendition.bitrate / 1e6;
			return `${renditionText} (${`${mbps.toFixed(mbps < 1 ? 1 : 0)} Mbps`})`;
		}
		return this.formatMenuItemText(renditionText, rendition);
	}
	static compareRendition(a, b) {
		var _a, _b;
		return b.height === a.height ? ((_a = b.bitrate) != null ? _a : 0) - ((_b = a.bitrate) != null ? _b : 0) : b.height - a.height;
	}
	attributeChangedCallback(attrName, oldValue, newValue) {
		super.attributeChangedCallback(attrName, oldValue, newValue);
		if (oldValue !== newValue) switch (attrName) {
			case MediaUIAttributes.MEDIA_RENDITION_SELECTED:
				this.value = newValue != null ? newValue : "auto";
				__privateMethod(this, _render, render_fn).call(this);
				break;
			case MediaUIAttributes.MEDIA_RENDITION_LIST:
				__privateSet(this, _renditionList, parseRenditionList(newValue));
				__privateMethod(this, _render, render_fn).call(this);
				break;
			case MediaUIAttributes.MEDIA_HEIGHT:
			case MediaUIAttributes.MEDIA_WIDTH:
				__privateMethod(this, _render, render_fn).call(this);
				break;
		}
	}
	connectedCallback() {
		super.connectedCallback();
		this.addEventListener("change", __privateMethod(this, _onChange, onChange_fn));
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.removeEventListener("change", __privateMethod(this, _onChange, onChange_fn));
	}
	/**
	* Returns the anchor element when it is a floating menu.
	*/
	get anchorElement() {
		if (this.anchor !== "auto") return super.anchorElement;
		return getMediaController(this).querySelector("media-rendition-menu-button");
	}
	get mediaRenditionList() {
		return __privateGet(this, _renditionList);
	}
	set mediaRenditionList(list) {
		__privateSet(this, _renditionList, list);
		__privateMethod(this, _render, render_fn).call(this);
	}
	/**
	* Get selected rendition id.
	*/
	get mediaRenditionSelected() {
		return getStringAttr(this, MediaUIAttributes.MEDIA_RENDITION_SELECTED);
	}
	set mediaRenditionSelected(id) {
		setStringAttr(this, MediaUIAttributes.MEDIA_RENDITION_SELECTED, id);
	}
	get mediaHeight() {
		return getNumericAttr(this, MediaUIAttributes.MEDIA_HEIGHT);
	}
	set mediaHeight(height) {
		setNumericAttr(this, MediaUIAttributes.MEDIA_HEIGHT, height);
	}
	get mediaWidth() {
		return getNumericAttr(this, MediaUIAttributes.MEDIA_WIDTH);
	}
	set mediaWidth(width) {
		setNumericAttr(this, MediaUIAttributes.MEDIA_WIDTH, width);
	}
	compareRendition(a, b) {
		return this.constructor.compareRendition(a, b);
	}
	formatMenuItemText(text, rendition) {
		return this.constructor.formatMenuItemText(text, rendition);
	}
	formatRendition(rendition, options) {
		return this.constructor.formatRendition(rendition, options);
	}
	showRenditionBitrate(rendition) {
		return this.mediaRenditionList.some((r) => r !== rendition && r.height === rendition.height && r.bitrate !== rendition.bitrate);
	}
};
_renditionList = /* @__PURE__ */ new WeakMap();
_prevState = /* @__PURE__ */ new WeakMap();
_render = /* @__PURE__ */ new WeakSet();
render_fn = function() {
	const isAuto = !this.mediaRenditionSelected;
	if (__privateGet(this, _prevState).mediaRenditionList === JSON.stringify(this.mediaRenditionList) && __privateGet(this, _prevState).mediaHeight === this.mediaHeight && __privateGet(this, _prevState).mediaWidth === this.mediaWidth && __privateGet(this, _prevState).isAuto === isAuto) return;
	__privateGet(this, _prevState).mediaRenditionList = JSON.stringify(this.mediaRenditionList);
	__privateGet(this, _prevState).mediaHeight = this.mediaHeight;
	__privateGet(this, _prevState).mediaWidth = this.mediaWidth;
	__privateGet(this, _prevState).isAuto = isAuto;
	const renditionList = this.mediaRenditionList.sort(this.compareRendition.bind(this));
	const selectedRendition = renditionList.find((rendition) => rendition.id === this.mediaRenditionSelected);
	for (const rendition of renditionList) rendition.selected = rendition === selectedRendition;
	this.defaultSlot.textContent = "";
	for (const rendition of renditionList) {
		const item2 = createMenuItem({
			type: "radio",
			text: this.formatRendition(rendition, { showBitrate: this.showRenditionBitrate(rendition) }),
			value: `${rendition.id}`,
			checked: rendition.selected && !isAuto
		});
		item2.prepend(createIndicator(this, "checked-indicator"));
		this.defaultSlot.append(item2);
	}
	const showSelectedBitrate = selectedRendition && this.showRenditionBitrate(selectedRendition);
	let autoText = void 0;
	if (isAuto) {
		if (selectedRendition) autoText = this.formatMenuItemText(`${t("Auto")} \u2022 ${this.formatRendition(selectedRendition, { showBitrate: showSelectedBitrate })}`, selectedRendition);
		else if (this.mediaHeight > 0 && this.mediaWidth > 0) autoText = this.formatMenuItemText(`${t("Auto")} (${Math.min(this.mediaWidth, this.mediaHeight)}p)`);
	}
	if (!autoText) autoText = this.formatMenuItemText(t("Auto"));
	const item = createMenuItem({
		type: "radio",
		text: autoText,
		value: "auto",
		checked: isAuto
	});
	item.dataset.description = autoText;
	item.prepend(createIndicator(this, "checked-indicator"));
	this.defaultSlot.append(item);
};
_onChange = /* @__PURE__ */ new WeakSet();
onChange_fn = function() {
	if (this.value == null) return;
	const event = new GlobalThis.CustomEvent(MediaUIEvents.MEDIA_RENDITION_REQUEST, {
		composed: true,
		bubbles: true,
		detail: this.value
	});
	this.dispatchEvent(event);
};
if (!GlobalThis.customElements.get("media-rendition-menu")) GlobalThis.customElements.define("media-rendition-menu", MediaRenditionMenu);
//#endregion
export { MediaRenditionMenu };
