// castable-video/castable-utils.js
//#region node_modules/castable-video/castable-utils.js
var privateProps = /* @__PURE__ */ new WeakMap();
var InvalidStateError = class extends Error {};
var NotSupportedError = class extends Error {};
var HLS_RESPONSE_HEADERS = [
	"application/x-mpegURL",
	"application/vnd.apple.mpegurl",
	"audio/mpegurl"
];
var IterableWeakSet = globalThis.WeakRef ? class extends Set {
	add(el) {
		super.add(new WeakRef(el));
	}
	forEach(fn) {
		super.forEach((ref) => {
			const value = ref.deref();
			if (value) fn(value);
		});
	}
} : Set;
function onCastApiAvailable(callback) {
	if (!globalThis.chrome?.cast?.isAvailable) globalThis.__onGCastApiAvailable = () => {
		customElements.whenDefined("google-cast-button").then(callback);
	};
	else if (!globalThis.cast?.framework) customElements.whenDefined("google-cast-button").then(callback);
	else callback();
}
function requiresCastFramework() {
	return globalThis.chrome;
}
function loadCastFramework() {
	const sdkUrl = "https://www.gstatic.com/cv/js/sender/v1/cast_sender.js?loadCastFramework=1";
	if (globalThis.chrome?.cast || document.querySelector(`script[src="${sdkUrl}"]`)) return;
	const script = document.createElement("script");
	script.src = sdkUrl;
	document.head.append(script);
}
function castContext() {
	return globalThis.cast?.framework?.CastContext.getInstance();
}
function currentSession() {
	return castContext()?.getCurrentSession();
}
function currentMedia() {
	return currentSession()?.getSessionObj().media[0];
}
function editTracksInfo(request) {
	return new Promise((resolve, reject) => {
		currentMedia().editTracksInfo(request, resolve, reject);
	});
}
function getMediaStatus(request) {
	return new Promise((resolve, reject) => {
		currentMedia().getStatus(request, resolve, reject);
	});
}
function setCastOptions(options) {
	return castContext().setOptions({
		...getDefaultCastOptions(),
		...options
	});
}
function getDefaultCastOptions() {
	return {
		receiverApplicationId: "CC1AD845",
		autoJoinPolicy: "origin_scoped",
		androidReceiverCompatible: false,
		language: "en-US",
		resumeSavedSession: true
	};
}
function getFormat(segment) {
	if (!segment) return void 0;
	const match = segment.match(/\.([a-zA-Z0-9]+)(?:\?.*)?$/);
	return match ? match[1] : null;
}
function parsePlaylistUrls(playlistContent) {
	const lines = playlistContent.split("\n");
	const urls = [];
	for (let i = 0; i < lines.length; i++) if (lines[i].trim().startsWith("#EXT-X-STREAM-INF")) {
		const nextLine = lines[i + 1] ? lines[i + 1].trim() : "";
		if (nextLine && !nextLine.startsWith("#")) urls.push(nextLine);
	}
	return urls;
}
function parseSegment(playlistContent) {
	return playlistContent.split("\n").find((line) => !line.trim().startsWith("#") && line.trim() !== "");
}
async function isHls(url) {
	if (!url) return false;
	if (/\.m3u8?(\?.*)?$/i.test(url)) return true;
	if (url.startsWith("blob:")) return false;
	try {
		const contentType = (await fetch(url, { method: "HEAD" })).headers.get("Content-Type");
		return HLS_RESPONSE_HEADERS.some((header) => contentType === header);
	} catch (err) {
		console.error("Error while trying to get the Content-Type of the manifest", err);
		return false;
	}
}
async function getPlaylistSegmentFormat(url) {
	if (!url || url.startsWith("blob:")) return void 0;
	try {
		const mainManifestContent = await (await fetch(url)).text();
		let availableChunksContent = mainManifestContent;
		const playlists = parsePlaylistUrls(mainManifestContent);
		if (playlists.length > 0) {
			const chosenPlaylistUrl = new URL(playlists[0], url).toString();
			availableChunksContent = await (await fetch(chosenPlaylistUrl)).text();
		}
		return getFormat(parseSegment(availableChunksContent));
	} catch (err) {
		console.error("Error while trying to parse the manifest playlist", err);
		return;
	}
}
//#endregion
export { InvalidStateError, IterableWeakSet, NotSupportedError, castContext, currentMedia, currentSession, editTracksInfo, getDefaultCastOptions, getMediaStatus, getPlaylistSegmentFormat, isHls, loadCastFramework, onCastApiAvailable, privateProps, requiresCastFramework, setCastOptions };
