// @supabase/auth-js/dist/module/lib/locks.js
import { supportsLocalStorage } from "./helpers.js";
//#region node_modules/@supabase/auth-js/dist/module/lib/locks.js
/**
* @experimental
*/
var internals = { 
/**
* @experimental
*/
debug: !!(globalThis && supportsLocalStorage() && globalThis.localStorage && globalThis.localStorage.getItem("supabase.gotrue-js.locks.debug") === "true") };
/**
* An error thrown when a lock cannot be acquired after some amount of time.
*
* Use the {@link #isAcquireTimeout} property instead of checking with `instanceof`.
*
* @example
* ```ts
* import { LockAcquireTimeoutError } from '@supabase/auth-js'
*
* class CustomLockError extends LockAcquireTimeoutError {
*   constructor() {
*     super('Lock timed out')
*   }
* }
* ```
*/
var LockAcquireTimeoutError = class extends Error {
	constructor(message) {
		super(message);
		this.isAcquireTimeout = true;
	}
};
/**
* Error thrown when the browser Navigator Lock API fails to acquire a lock.
*
* @example
* ```ts
* import { NavigatorLockAcquireTimeoutError } from '@supabase/auth-js'
*
* throw new NavigatorLockAcquireTimeoutError('Lock timed out')
* ```
*/
var NavigatorLockAcquireTimeoutError = class extends LockAcquireTimeoutError {};
/**
* Implements a global exclusive lock using the Navigator LockManager API. It
* is available on all browsers released after 2022-03-15 with Safari being the
* last one to release support. If the API is not available, this function will
* throw. Make sure you check availablility before configuring {@link
* GoTrueClient}.
*
* You can turn on debugging by setting the `supabase.gotrue-js.locks.debug`
* local storage item to `true`.
*
* Internals:
*
* Since the LockManager API does not preserve stack traces for the async
* function passed in the `request` method, a trick is used where acquiring the
* lock releases a previously started promise to run the operation in the `fn`
* function. The lock waits for that promise to finish (with or without error),
* while the function will finally wait for the result anyway.
*
* @param name Name of the lock to be acquired.
* @param acquireTimeout If negative, no timeout. If 0 an error is thrown if
*                       the lock can't be acquired without waiting. If positive, the lock acquire
*                       will time out after so many milliseconds. An error is
*                       a timeout if it has `isAcquireTimeout` set to true.
* @param fn The operation to run once the lock is acquired.
* @example
* ```ts
* await navigatorLock('sync-user', 1000, async () => {
*   await refreshSession()
* })
* ```
*/
async function navigatorLock(name, acquireTimeout, fn) {
	if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: acquire lock", name, acquireTimeout);
	const abortController = new globalThis.AbortController();
	let acquireTimeoutTimer;
	if (acquireTimeout > 0) acquireTimeoutTimer = setTimeout(() => {
		abortController.abort();
		if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock acquire timed out", name);
	}, acquireTimeout);
	await Promise.resolve();
	try {
		return await globalThis.navigator.locks.request(name, acquireTimeout === 0 ? {
			mode: "exclusive",
			ifAvailable: true
		} : {
			mode: "exclusive",
			signal: abortController.signal
		}, async (lock) => {
			if (lock) {
				clearTimeout(acquireTimeoutTimer);
				if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: acquired", name, lock.name);
				try {
					return await fn();
				} finally {
					if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: released", name, lock.name);
				}
			} else if (acquireTimeout === 0) {
				if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: not immediately available", name);
				throw new NavigatorLockAcquireTimeoutError(`Acquiring an exclusive Navigator LockManager lock "${name}" immediately failed`);
			} else {
				if (internals.debug) try {
					const result = await globalThis.navigator.locks.query();
					console.log("@supabase/gotrue-js: Navigator LockManager state", JSON.stringify(result, null, "  "));
				} catch (e) {
					console.warn("@supabase/gotrue-js: Error when querying Navigator LockManager state", e);
				}
				console.warn("@supabase/gotrue-js: Navigator LockManager returned a null lock when using #request without ifAvailable set to true, it appears this browser is not following the LockManager spec https://developer.mozilla.org/en-US/docs/Web/API/LockManager/request");
				clearTimeout(acquireTimeoutTimer);
				return await fn();
			}
		});
	} catch (e) {
		if (acquireTimeout > 0) clearTimeout(acquireTimeoutTimer);
		if ((e === null || e === void 0 ? void 0 : e.name) === "AbortError" && acquireTimeout > 0) if (abortController.signal.aborted) {
			if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: acquire timeout, recovering by stealing lock", name);
			console.warn(`@supabase/gotrue-js: Lock "${name}" was not released within ${acquireTimeout}ms. This may indicate an orphaned lock from a component unmount (e.g., React Strict Mode). Forcefully acquiring the lock to recover.`);
			return await Promise.resolve().then(() => globalThis.navigator.locks.request(name, {
				mode: "exclusive",
				steal: true
			}, async (lock) => {
				if (lock) {
					if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: recovered (stolen)", name, lock.name);
					try {
						return await fn();
					} finally {
						if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: released (stolen)", name, lock.name);
					}
				} else {
					console.warn("@supabase/gotrue-js: Navigator LockManager returned null lock even with steal: true");
					return await fn();
				}
			}));
		} else {
			if (internals.debug) console.log("@supabase/gotrue-js: navigatorLock: lock was stolen by another request", name);
			throw new NavigatorLockAcquireTimeoutError(`Lock "${name}" was released because another request stole it`);
		}
		throw e;
	}
}
//#endregion
export { LockAcquireTimeoutError, NavigatorLockAcquireTimeoutError, internals, navigatorLock };
