// @journeyapps/wa-sqlite/src/VFS.js
import { SQLITE_OPEN_MAIN_JOURNAL, SQLITE_OPEN_SUBJOURNAL, SQLITE_OPEN_SUPER_JOURNAL, SQLITE_OPEN_TEMP_JOURNAL, SQLITE_OPEN_TRANSIENT_DB, SQLITE_OPEN_WAL } from "./sqlite-constants.js";
//#region node_modules/@journeyapps/wa-sqlite/src/VFS.js
var DEFAULT_SECTOR_SIZE = 512;
var Base = class {
	name;
	mxPathname = 64;
	_module;
	/**
	* @param {string} name 
	* @param {object} module 
	*/
	constructor(name, module) {
		this.name = name;
		this._module = module;
	}
	/**
	* @returns {void|Promise<void>} 
	*/
	close() {}
	/**
	* @returns {boolean|Promise<boolean>}
	*/
	isReady() {
		return true;
	}
	/**
	* Overload in subclasses to indicate which methods are asynchronous.
	* @param {string} methodName 
	* @returns {boolean}
	*/
	hasAsyncMethod(methodName) {
		return false;
	}
	/**
	* @param {number} pVfs 
	* @param {number} zName 
	* @param {number} pFile 
	* @param {number} flags 
	* @param {number} pOutFlags 
	* @returns {number|Promise<number>}
	*/
	xOpen(pVfs, zName, pFile, flags, pOutFlags) {
		return 14;
	}
	/**
	* @param {number} pVfs 
	* @param {number} zName 
	* @param {number} syncDir 
	* @returns {number|Promise<number>}
	*/
	xDelete(pVfs, zName, syncDir) {
		return 0;
	}
	/**
	* @param {number} pVfs 
	* @param {number} zName 
	* @param {number} flags 
	* @param {number} pResOut 
	* @returns {number|Promise<number>}
	*/
	xAccess(pVfs, zName, flags, pResOut) {
		return 0;
	}
	/**
	* @param {number} pVfs 
	* @param {number} zName 
	* @param {number} nOut 
	* @param {number} zOut 
	* @returns {number|Promise<number>}
	*/
	xFullPathname(pVfs, zName, nOut, zOut) {
		return 0;
	}
	/**
	* @param {number} pVfs 
	* @param {number} nBuf 
	* @param {number} zBuf 
	* @returns {number|Promise<number>}
	*/
	xGetLastError(pVfs, nBuf, zBuf) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @returns {number|Promise<number>}
	*/
	xClose(pFile) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} pData 
	* @param {number} iAmt 
	* @param {number} iOffsetLo 
	* @param {number} iOffsetHi 
	* @returns {number|Promise<number>}
	*/
	xRead(pFile, pData, iAmt, iOffsetLo, iOffsetHi) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} pData 
	* @param {number} iAmt 
	* @param {number} iOffsetLo 
	* @param {number} iOffsetHi 
	* @returns {number|Promise<number>}
	*/
	xWrite(pFile, pData, iAmt, iOffsetLo, iOffsetHi) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} sizeLo 
	* @param {number} sizeHi 
	* @returns {number|Promise<number>}
	*/
	xTruncate(pFile, sizeLo, sizeHi) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} flags 
	* @returns {number|Promise<number>}
	*/
	xSync(pFile, flags) {
		return 0;
	}
	/**
	* 
	* @param {number} pFile 
	* @param {number} pSize 
	* @returns {number|Promise<number>}
	*/
	xFileSize(pFile, pSize) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} lockType 
	* @returns {number|Promise<number>}
	*/
	xLock(pFile, lockType) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} lockType 
	* @returns {number|Promise<number>}
	*/
	xUnlock(pFile, lockType) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} pResOut 
	* @returns {number|Promise<number>}
	*/
	xCheckReservedLock(pFile, pResOut) {
		return 0;
	}
	/**
	* @param {number} pFile 
	* @param {number} op 
	* @param {number} pArg 
	* @returns {number|Promise<number>}
	*/
	xFileControl(pFile, op, pArg) {
		return 12;
	}
	/**
	* @param {number} pFile 
	* @returns {number|Promise<number>}
	*/
	xSectorSize(pFile) {
		return DEFAULT_SECTOR_SIZE;
	}
	/**
	* @param {number} pFile 
	* @returns {number|Promise<number>}
	*/
	xDeviceCharacteristics(pFile) {
		return 0;
	}
};
[
	256,
	SQLITE_OPEN_MAIN_JOURNAL,
	512,
	SQLITE_OPEN_TEMP_JOURNAL,
	SQLITE_OPEN_TRANSIENT_DB,
	SQLITE_OPEN_SUBJOURNAL,
	SQLITE_OPEN_SUPER_JOURNAL,
	SQLITE_OPEN_WAL
].reduce((mask, element) => mask | element);
//#endregion
export { Base };
