// media-chrome/dist/menu/media-chrome-menu-button.js
import { GlobalThis } from "../utils/server-safe-globals.js";
import { getDocumentOrShadowRoot } from "../utils/element-utils.js";
import { MediaChromeButton } from "../media-chrome-button.js";
import { InvokeEvent } from "../utils/events.js";
//#region node_modules/media-chrome/dist/menu/media-chrome-menu-button.js
var MediaChromeMenuButton = class extends MediaChromeButton {
	connectedCallback() {
		super.connectedCallback();
		if (this.invokeTargetElement) this.setAttribute("aria-haspopup", "menu");
	}
	get invokeTarget() {
		return this.getAttribute("invoketarget");
	}
	set invokeTarget(value) {
		this.setAttribute("invoketarget", `${value}`);
	}
	/**
	* Returns the element with the id specified by the `invoketarget` attribute.
	* @return {HTMLElement | null}
	*/
	get invokeTargetElement() {
		var _a;
		if (this.invokeTarget) return (_a = getDocumentOrShadowRoot(this)) == null ? void 0 : _a.querySelector(`#${this.invokeTarget}`);
		return null;
	}
	handleClick() {
		var _a;
		(_a = this.invokeTargetElement) == null || _a.dispatchEvent(new InvokeEvent({ relatedTarget: this }));
	}
};
if (!GlobalThis.customElements.get("media-chrome-menu-button")) GlobalThis.customElements.define("media-chrome-menu-button", MediaChromeMenuButton);
//#endregion
export { MediaChromeMenuButton };
