// @journeyapps/wa-sqlite/src/WebLocksMixin.js
import { SQLITE_IOERR_CHECKRESERVEDLOCK, SQLITE_IOERR_LOCK, SQLITE_IOERR_UNLOCK } from "./sqlite-constants.js";
import "./VFS.js";
//#region node_modules/@journeyapps/wa-sqlite/src/WebLocksMixin.js
/** @type {LockOptions} */ var SHARED = { mode: "shared" };
/** @type {LockOptions} */ var POLL_SHARED = {
	ifAvailable: true,
	mode: "shared"
};
/** @type {LockOptions} */ var POLL_EXCLUSIVE = {
	ifAvailable: true,
	mode: "exclusive"
};
var POLICIES = [
	"exclusive",
	"shared",
	"shared+hint"
];
/**
* @typedef LockState
* @property {string} baseName
* @property {number} type
* @property {boolean} writeHint
* 
* These properties are functions that release a specific lock.
* @property {(() => void)?} [gate]
* @property {(() => void)?} [access]
* @property {(() => void)?} [reserved]
* @property {(() => void)?} [hint]
*/
/**
* Mix-in for FacadeVFS that implements the SQLite VFS locking protocol.
* @param {*} superclass FacadeVFS (or subclass)
* @returns 
*/
var WebLocksMixin = (superclass) => class extends superclass {
	#options = {
		lockPolicy: "exclusive",
		lockTimeout: Infinity
	};
	/** @type {Map<number, LockState>} */ #mapIdToState = /* @__PURE__ */ new Map();
	constructor(name, module, options) {
		super(name, module, options);
		Object.assign(this.#options, options);
		if (POLICIES.indexOf(this.#options.lockPolicy) === -1) throw new Error(`WebLocksMixin: invalid lock mode: ${options.lockPolicy}`);
	}
	/**
	* @param {number} fileId 
	* @param {number} lockType 
	* @returns {Promise<number>}
	*/
	async jLock(fileId, lockType) {
		try {
			const lockState = this.#getLockState(fileId);
			if (lockType <= lockState.type) return 0;
			switch (this.#options.lockPolicy) {
				case "exclusive": return await this.#lockExclusive(lockState, lockType);
				case "shared":
				case "shared+hint": return await this.#lockShared(lockState, lockType);
			}
		} catch (e) {
			console.error("WebLocksMixin: lock error", e);
			return SQLITE_IOERR_LOCK;
		}
	}
	/**
	* @param {number} fileId 
	* @param {number} lockType 
	* @returns {Promise<number>}
	*/
	async jUnlock(fileId, lockType) {
		try {
			const lockState = this.#getLockState(fileId);
			if (!(lockType < lockState.type)) return 0;
			switch (this.#options.lockPolicy) {
				case "exclusive": return await this.#unlockExclusive(lockState, lockType);
				case "shared":
				case "shared+hint": return await this.#unlockShared(lockState, lockType);
			}
		} catch (e) {
			console.error("WebLocksMixin: unlock error", e);
			return SQLITE_IOERR_UNLOCK;
		}
	}
	/**
	* @param {number} fileId 
	* @param {DataView} pResOut 
	* @returns {Promise<number>}
	*/
	async jCheckReservedLock(fileId, pResOut) {
		try {
			const lockState = this.#getLockState(fileId);
			switch (this.#options.lockPolicy) {
				case "exclusive": return this.#checkReservedExclusive(lockState, pResOut);
				case "shared":
				case "shared+hint": return await this.#checkReservedShared(lockState, pResOut);
			}
		} catch (e) {
			console.error("WebLocksMixin: check reserved lock error", e);
			return SQLITE_IOERR_CHECKRESERVEDLOCK;
		}
		pResOut.setInt32(0, 0, true);
		return 0;
	}
	/**
	* @param {number} fileId
	* @param {number} op
	* @param {DataView} pArg
	* @returns {number|Promise<number>}
	*/
	jFileControl(fileId, op, pArg) {
		if (op === WebLocksMixin.WRITE_HINT_OP_CODE && this.#options.lockPolicy === "shared+hint") {
			const lockState = this.#getLockState(fileId);
			lockState.writeHint = true;
		}
		return 12;
	}
	#getLockState(fileId) {
		let lockState = this.#mapIdToState.get(fileId);
		if (!lockState) {
			lockState = {
				baseName: this.getFilename(fileId),
				type: 0,
				writeHint: false
			};
			this.#mapIdToState.set(fileId, lockState);
		}
		return lockState;
	}
	/**
	* @param {LockState} lockState 
	* @param {number} lockType 
	* @returns 
	*/
	async #lockExclusive(lockState, lockType) {
		if (!lockState.access) {
			if (!await this.#acquire(lockState, "access")) return 5;
			console.assert(!!lockState.access);
		}
		lockState.type = lockType;
		return 0;
	}
	/**
	* @param {LockState} lockState 
	* @param {number} lockType 
	* @returns {number}
	*/
	#unlockExclusive(lockState, lockType) {
		if (lockType === 0) {
			lockState.access?.();
			console.assert(!lockState.access);
		}
		lockState.type = lockType;
		return 0;
	}
	/**
	* @param {LockState} lockState 
	* @param {DataView} pResOut 
	* @returns {number}
	*/
	#checkReservedExclusive(lockState, pResOut) {
		pResOut.setInt32(0, 0, true);
		return 0;
	}
	/**
	* @param {LockState} lockState 
	* @param {number} lockType 
	* @returns 
	*/
	async #lockShared(lockState, lockType) {
		switch (lockState.type) {
			case 0:
				switch (lockType) {
					case 1:
						if (lockState.writeHint) {
							if (!await this.#acquire(lockState, "hint")) return 5;
						}
						if (!await this.#acquire(lockState, "gate", SHARED)) {
							lockState.hint?.();
							return 5;
						}
						await this.#acquire(lockState, "access", SHARED);
						lockState.gate();
						console.assert(!lockState.gate);
						console.assert(!!lockState.access);
						console.assert(!lockState.reserved);
						break;
					default: throw new Error("unsupported lock transition");
				}
				break;
			case 1:
				switch (lockType) {
					case 2:
						if (this.#options.lockPolicy === "shared+hint") {
							if (!lockState.hint && !await this.#acquire(lockState, "hint", POLL_EXCLUSIVE)) return 5;
						}
						if (!await this.#acquire(lockState, "reserved", POLL_EXCLUSIVE)) {
							lockState.hint?.();
							return 5;
						}
						lockState.access();
						console.assert(!lockState.gate);
						console.assert(!lockState.access);
						console.assert(!!lockState.reserved);
						break;
					case 4:
						if (!await this.#acquire(lockState, "gate")) return 5;
						lockState.access();
						if (!await this.#acquire(lockState, "access")) {
							lockState.gate();
							return 5;
						}
						console.assert(!!lockState.gate);
						console.assert(!!lockState.access);
						console.assert(!lockState.reserved);
						break;
					default: throw new Error("unsupported lock transition");
				}
				break;
			case 2:
				switch (lockType) {
					case 4:
						if (!await this.#acquire(lockState, "gate")) return 5;
						if (!await this.#acquire(lockState, "access")) {
							lockState.gate();
							return 5;
						}
						console.assert(!!lockState.gate);
						console.assert(!!lockState.access);
						console.assert(!!lockState.reserved);
						break;
					default: throw new Error("unsupported lock transition");
				}
				break;
		}
		lockState.type = lockType;
		return 0;
	}
	/**
	* @param {LockState} lockState 
	* @param {number} lockType 
	* @returns 
	*/
	async #unlockShared(lockState, lockType) {
		if (lockType === 0) {
			lockState.access?.();
			lockState.gate?.();
			lockState.reserved?.();
			lockState.hint?.();
			lockState.writeHint = false;
			console.assert(!lockState.access);
			console.assert(!lockState.gate);
			console.assert(!lockState.reserved);
			console.assert(!lockState.hint);
		} else switch (lockState.type) {
			case 4:
				lockState.access();
				await this.#acquire(lockState, "access", SHARED);
				lockState.gate();
				lockState.reserved?.();
				lockState.hint?.();
				console.assert(!!lockState.access);
				console.assert(!lockState.gate);
				console.assert(!lockState.reserved);
				break;
			case 2:
				await this.#acquire(lockState, "access", SHARED);
				lockState.reserved();
				lockState.hint?.();
				console.assert(!!lockState.access);
				console.assert(!lockState.gate);
				console.assert(!lockState.reserved);
				break;
		}
		lockState.type = lockType;
		return 0;
	}
	/**
	* @param {LockState} lockState 
	* @param {DataView} pResOut 
	* @returns {Promise<number>}
	*/
	async #checkReservedShared(lockState, pResOut) {
		if (await this.#acquire(lockState, "reserved", POLL_SHARED)) {
			lockState.reserved();
			pResOut.setInt32(0, 0, true);
		} else pResOut.setInt32(0, 1, true);
		return 0;
	}
	/**
	* @param {LockState} lockState 
	* @param {'gate'|'access'|'reserved'|'hint'} name
	* @param {LockOptions} options 
	* @returns {Promise<boolean>}
	*/
	#acquire(lockState, name, options = {}) {
		console.assert(!lockState[name]);
		return new Promise((resolve) => {
			if (!options.ifAvailable && this.#options.lockTimeout < Infinity) {
				const controller = new AbortController();
				options = Object.assign({}, options, { signal: controller.signal });
				setTimeout(() => {
					controller.abort();
					resolve?.(false);
				}, this.#options.lockTimeout);
			}
			const lockName = `lock##${lockState.baseName}##${name}`;
			navigator.locks.request(lockName, options, (lock) => {
				if (lock) return new Promise((release) => {
					lockState[name] = () => {
						release();
						lockState[name] = null;
					};
					resolve(true);
					resolve = null;
				});
				else {
					lockState[name] = null;
					resolve(false);
					resolve = null;
				}
			}).catch((e) => {
				if (e.name !== "AbortError") throw e;
			});
		});
	}
};
WebLocksMixin.WRITE_HINT_OP_CODE = -9999;
//#endregion
export { WebLocksMixin };
