// @powersync/web/lib/src/db/adapters/AsyncWebAdapter.js
import { DBAdapterDefaultMixin, Mutex, Semaphore } from "../../../../../common/dist/bundle.js";
//#region node_modules/@powersync/web/lib/src/db/adapters/AsyncWebAdapter.js
/**
* A connection pool implementation delegating to another pool opened asynchronnously.
*/
var AsyncConnectionPool = class {
	name;
	state;
	resolvedWriter;
	pendingListeners = /* @__PURE__ */ new Set();
	constructor(inner, name) {
		this.name = name;
		this.state = inner.then((client) => {
			for (const pending of this.pendingListeners) pending.closeAfterRegisteredOnResolvedPool = client.writer.registerListener(pending.listener);
			this.pendingListeners.clear();
			this.resolvedWriter = client.writer;
			if (client.additionalReaders.length) return readWritePoolState(client.writer, client.additionalReaders);
			return singleConnectionPoolState(client.writer);
		});
	}
	async init() {
		await this.state;
	}
	async close() {
		await (await this.state).close();
	}
	async readLock(fn, options) {
		return (await this.state).withConnection(true, fn, options);
	}
	async writeLock(fn, options) {
		return (await this.state).withConnection(false, fn, options);
	}
	async refreshSchema() {
		await (await this.state).refreshSchema();
	}
	registerListener(listener) {
		if (this.resolvedWriter) return this.resolvedWriter.registerListener(listener);
		else {
			const pending = { listener };
			this.pendingListeners.add(pending);
			return () => {
				if (pending.closeAfterRegisteredOnResolvedPool) return pending.closeAfterRegisteredOnResolvedPool();
				else this.pendingListeners.delete(pending);
			};
		}
	}
};
function singleConnectionPoolState(connection) {
	return {
		writer: connection,
		withConnection: (allowReadOnly, fn, options) => {
			if (allowReadOnly) return connection.readLock(fn, options);
			else return connection.writeLock(fn, options);
		},
		close: () => connection.close(),
		refreshSchema: () => connection.refreshSchema()
	};
}
function readWritePoolState(writer, readers) {
	const writerMutex = new Mutex();
	const readerSemaphore = new Semaphore(readers);
	return {
		writer,
		async withConnection(allowReadOnly, fn, options) {
			const abortController = new AbortController();
			const abortSignal = abortController.signal;
			let timeout = null;
			let release;
			if (options?.timeoutMs) timeout = setTimeout(() => abortController.abort, options.timeoutMs);
			try {
				if (allowReadOnly) {
					let connection;
					[connection, release] = await new Promise((resolve, reject) => {
						let didComplete = false;
						function complete() {
							didComplete = true;
							abortController.abort();
						}
						function completeSuccess(connection, returnFn) {
							if (didComplete) returnFn();
							else {
								complete();
								resolve([connection, returnFn]);
							}
						}
						function completeError(error) {
							if (didComplete) return;
							complete();
							reject(error);
						}
						writerMutex.acquire(abortSignal).then((unlock) => completeSuccess(writer, unlock), completeError);
						readerSemaphore.requestOne(abortSignal).then(({ item, release }) => completeSuccess(item, release), completeError);
					});
					return await connection.readLock(fn);
				} else return await writerMutex.runExclusive(() => writer.writeLock(fn), abortSignal);
			} finally {
				if (timeout != null) clearTimeout(timeout);
				release?.();
			}
		},
		async close() {
			await writer.close();
			await Promise.all(readers.map((r) => r.close()));
		},
		async refreshSchema() {
			await writer.refreshSchema();
			await Promise.all(readers.map((r) => r.refreshSchema()));
		}
	};
}
var AsyncDbAdapter = class extends DBAdapterDefaultMixin(AsyncConnectionPool) {
	async shareConnection() {
		return (await this.state).writer.shareConnection();
	}
	getConfiguration() {
		if (this.resolvedWriter) return this.resolvedWriter.getConfiguration();
		throw new Error("AsyncDbAdapter.getConfiguration() can only be called after initializing it.");
	}
};
//#endregion
export { AsyncDbAdapter };
