// @powersync/web/lib/src/worker/db/MultiDatabaseServer.js
import { getNavigatorLocks } from "../../shared/navigator.js";
import { proxy } from "../../../../../../comlink/dist/esm/comlink.js";
import { DatabaseServer } from "../../db/adapters/wa-sqlite/DatabaseServer.js";
import { RawSqliteConnection } from "../../db/adapters/wa-sqlite/RawSqliteConnection.js";
import { ConcurrentSqliteConnection } from "../../db/adapters/wa-sqlite/ConcurrentConnection.js";
//#region node_modules/@powersync/web/lib/src/worker/db/MultiDatabaseServer.js
var OPEN_DB_LOCK = "open-wasqlite-db";
/**
* Shared state to manage multiple database connections hosted by a worker.
*/
var MultiDatabaseServer = class {
	logger;
	activeDatabases = /* @__PURE__ */ new Map();
	constructor(logger) {
		this.logger = logger;
	}
	async handleConnection(options) {
		this.logger.setLevel(options.logLevel);
		return proxy(await this.openConnectionLocally(options, options.lockName));
	}
	async connectToExisting(name, lockName) {
		return getNavigatorLocks().request(OPEN_DB_LOCK, async () => {
			const server = this.activeDatabases.get(name);
			if (server == null) throw new Error(`connectToExisting(${name}) failed because the worker doesn't own a database with that name.`);
			return proxy(await server.connect(lockName));
		});
	}
	async openConnectionLocally(options, lockName) {
		const maxAttempts = 3;
		let server;
		for (let count = 0; count < maxAttempts - 1; count++) try {
			server = await this.databaseOpenAttempt(options);
		} catch (ex) {
			this.logger.warn(`Attempt ${count + 1} of ${maxAttempts} to open database failed, retrying in 1 second...`, ex);
			await new Promise((resolve) => setTimeout(resolve, 1e3));
		}
		server ??= await this.databaseOpenAttempt(options);
		return server.connect(lockName);
	}
	async databaseOpenAttempt(options) {
		return getNavigatorLocks().request(OPEN_DB_LOCK, async () => {
			const { dbFilename } = options;
			let server = this.activeDatabases.get(dbFilename);
			if (server == null) {
				const needsNavigatorLocks = !(isSharedWorker || options.isReadOnly);
				const connection = new RawSqliteConnection(options);
				const withSafeConcurrency = new ConcurrentSqliteConnection(connection, needsNavigatorLocks);
				const returnLease = await withSafeConcurrency.acquireMutex();
				try {
					await connection.init();
				} catch (e) {
					returnLease();
					await connection.close();
					throw e;
				}
				returnLease();
				const onClose = () => this.activeDatabases.delete(dbFilename);
				server = new DatabaseServer({
					inner: withSafeConcurrency,
					logger: this.logger,
					onClose
				});
				this.activeDatabases.set(dbFilename, server);
			}
			return server;
		});
	}
	closeAll() {
		const existingDatabases = [...this.activeDatabases.values()];
		return Promise.all(existingDatabases.map((db) => {
			db.forceClose();
		}));
	}
};
var isSharedWorker = "SharedWorkerGlobalScope" in globalThis;
//#endregion
export { MultiDatabaseServer };
