// chrono-node/dist/esm/common/casualReferences.js
import { Meridiem } from "../types.js";
import { assignSimilarDate, assignSimilarTime, implySimilarTime } from "../utils/dates.js";
import { ParsingComponents } from "../results.js";
//#region node_modules/chrono-node/dist/esm/common/casualReferences.js
function now(reference) {
	const targetDate = reference.getDateWithAdjustedTimezone();
	const component = new ParsingComponents(reference, {});
	assignSimilarDate(component, targetDate);
	assignSimilarTime(component, targetDate);
	component.assign("timezoneOffset", reference.getTimezoneOffset());
	component.addTag("casualReference/now");
	return component;
}
function today(reference) {
	const targetDate = reference.getDateWithAdjustedTimezone();
	const component = new ParsingComponents(reference, {});
	assignSimilarDate(component, targetDate);
	implySimilarTime(component, targetDate);
	component.delete("meridiem");
	component.addTag("casualReference/today");
	return component;
}
function yesterday(reference) {
	return theDayBefore(reference, 1).addTag("casualReference/yesterday");
}
function tomorrow(reference) {
	return theDayAfter(reference, 1).addTag("casualReference/tomorrow");
}
function theDayBefore(reference, numDay) {
	return theDayAfter(reference, -numDay);
}
function theDayAfter(reference, nDays) {
	const targetDate = reference.getDateWithAdjustedTimezone();
	const component = new ParsingComponents(reference, {});
	const newDate = new Date(targetDate.getTime());
	newDate.setDate(newDate.getDate() + nDays);
	assignSimilarDate(component, newDate);
	implySimilarTime(component, newDate);
	component.delete("meridiem");
	return component;
}
function tonight(reference, implyHour = 22) {
	const targetDate = reference.getDateWithAdjustedTimezone();
	const component = new ParsingComponents(reference, {});
	assignSimilarDate(component, targetDate);
	component.imply("hour", implyHour);
	component.imply("meridiem", Meridiem.PM);
	component.addTag("casualReference/tonight");
	return component;
}
function evening(reference, implyHour = 20) {
	const component = new ParsingComponents(reference, {});
	component.imply("meridiem", Meridiem.PM);
	component.imply("hour", implyHour);
	component.addTag("casualReference/evening");
	return component;
}
function midnight(reference) {
	const component = new ParsingComponents(reference, {});
	if (reference.getDateWithAdjustedTimezone().getHours() > 2) component.addDurationAsImplied({ day: 1 });
	component.assign("hour", 0);
	component.imply("minute", 0);
	component.imply("second", 0);
	component.imply("millisecond", 0);
	component.addTag("casualReference/midnight");
	return component;
}
function morning(reference, implyHour = 6) {
	const component = new ParsingComponents(reference, {});
	component.imply("meridiem", Meridiem.AM);
	component.imply("hour", implyHour);
	component.imply("minute", 0);
	component.imply("second", 0);
	component.imply("millisecond", 0);
	component.addTag("casualReference/morning");
	return component;
}
function afternoon(reference, implyHour = 15) {
	const component = new ParsingComponents(reference, {});
	component.imply("meridiem", Meridiem.PM);
	component.imply("hour", implyHour);
	component.imply("minute", 0);
	component.imply("second", 0);
	component.imply("millisecond", 0);
	component.addTag("casualReference/afternoon");
	return component;
}
function noon(reference) {
	const component = new ParsingComponents(reference, {});
	component.imply("meridiem", Meridiem.AM);
	component.assign("hour", 12);
	component.imply("minute", 0);
	component.imply("second", 0);
	component.imply("millisecond", 0);
	component.addTag("casualReference/noon");
	return component;
}
//#endregion
export { afternoon, evening, midnight, morning, noon, now, theDayAfter, today, tomorrow, tonight, yesterday };
