// media-chrome/dist/menu/media-context-menu-item.js
import { GlobalThis } from "../utils/server-safe-globals.js";
import { MediaChromeMenuItem } from "./media-chrome-menu-item.js";
//#region node_modules/media-chrome/dist/menu/media-context-menu-item.js
function getTemplateHTML(_attrs) {
	return `
    ${MediaChromeMenuItem.getTemplateHTML.call(this, _attrs)}
    <style>
        ::slotted(*) {
            color: var(--media-text-color, white);
            text-decoration: none;
            border: none;
            background: none;
            cursor: pointer;
            padding: 0;
            min-height: var(--media-control-height, 24px);
        }
    </style>
  `;
}
var MediaContextMenuItem = class extends MediaChromeMenuItem {};
MediaContextMenuItem.shadowRootOptions = { mode: "open" };
MediaContextMenuItem.getTemplateHTML = getTemplateHTML;
if (!GlobalThis.customElements.get("media-context-menu-item")) GlobalThis.customElements.define("media-context-menu-item", MediaContextMenuItem);
//#endregion
export { MediaContextMenuItem };
