// @journeyapps/wa-sqlite/src/examples/LazyLock.js
import { Lock } from "./Lock.js";
//#region node_modules/@journeyapps/wa-sqlite/src/examples/LazyLock.js
var LazyLock = class extends Lock {
	#channel;
	#isBusy = false;
	#hasReleaseRequest = false;
	/**
	* @param {string} name 
	*/
	constructor(name) {
		super(name);
		this.#channel = new BroadcastChannel(name);
		this.#channel.onmessage = (event) => {
			if (this.#isBusy) this.#hasReleaseRequest = true;
			else this.release();
		};
	}
	close() {
		super.close();
		this.#channel.onmessage = null;
		this.#channel.close();
	}
	/**
	* @param {LockMode} mode 
	* @param {number} timeout 
	* @returns {Promise<boolean>}
	*/
	async acquire(mode, timeout = -1) {
		this.#isBusy = true;
		try {
			if (mode === this.mode) return true;
			if (this.mode) super.release();
			else if (await super.acquire(mode, 0)) return true;
			const pResult = super.acquire(mode, timeout);
			this.#channel.postMessage({});
			return await pResult;
		} catch (e) {
			this.release();
			throw e;
		}
	}
	/**
	* @param {LockMode} mode 
	* @returns {boolean}
	*/
	acquireIfHeld(mode) {
		if (mode === this.mode) {
			this.#isBusy = true;
			return true;
		}
		return false;
	}
	release() {
		super.release();
		this.#isBusy = false;
		this.#hasReleaseRequest = false;
	}
	releaseLazy() {
		this.#isBusy = false;
		if (this.#hasReleaseRequest) this.release();
	}
};
//#endregion
export { LazyLock };
