// @powersync/web/lib/src/db/adapters/wa-sqlite/vfs.js
const __vite__mapDeps=(i,m=__vite__mapDeps,d=(m.f||(m.f=["node_modules/@journeyapps/wa-sqlite/src/examples/IDBBatchAtomicVFS.js","node_modules/@journeyapps/wa-sqlite/src/FacadeVFS.js","node_modules/@journeyapps/wa-sqlite/src/VFS.js","node_modules/@journeyapps/wa-sqlite/src/sqlite-constants.js","node_modules/@journeyapps/wa-sqlite/src/WebLocksMixin.js","node_modules/@journeyapps/wa-sqlite/src/examples/AccessHandlePoolVFS.js","node_modules/@journeyapps/wa-sqlite/src/examples/OPFSCoopSyncVFS.js","node_modules/@journeyapps/wa-sqlite/src/examples/OPFSWriteAheadVFS.js","node_modules/@journeyapps/wa-sqlite/src/examples/LazyLock.js","node_modules/@journeyapps/wa-sqlite/src/examples/Lock.js","node_modules/@journeyapps/wa-sqlite/src/examples/WriteAhead.js"])))=>i.map(i=>d[i]);
import { __vitePreload } from "../../../../../../../../_virtual/_vite/preload-helper.js";
//#region node_modules/@powersync/web/lib/src/db/adapters/wa-sqlite/vfs.js
/**
* List of currently tested virtual filesystems
*/
var WASQLiteVFS;
(function(WASQLiteVFS) {
	WASQLiteVFS["IDBBatchAtomicVFS"] = "IDBBatchAtomicVFS";
	WASQLiteVFS["OPFSCoopSyncVFS"] = "OPFSCoopSyncVFS";
	WASQLiteVFS["AccessHandlePoolVFS"] = "AccessHandlePoolVFS";
	WASQLiteVFS["OPFSWriteAheadVFS"] = "OPFSWriteAheadVFS";
})(WASQLiteVFS || (WASQLiteVFS = {}));
function vfsRequiresDedicatedWorkers(vfs) {
	return vfs != WASQLiteVFS.IDBBatchAtomicVFS;
}
async function asyncModuleFactory(encryptionKey) {
	if (encryptionKey) {
		const { default: factory } = await __vitePreload(async () => {
			const { default: factory } = await import("../../../../../../../@journeyapps/wa-sqlite/dist/mc-wa-sqlite-async.js");
			return { default: factory };
		}, []);
		return factory();
	} else {
		const { default: factory } = await __vitePreload(async () => {
			const { default: factory } = await import("../../../../../../../@journeyapps/wa-sqlite/dist/wa-sqlite-async.js");
			return { default: factory };
		}, []);
		return factory();
	}
}
async function syncModuleFactory(encryptionKey) {
	if (encryptionKey) {
		const { default: factory } = await __vitePreload(async () => {
			const { default: factory } = await import("../../../../../../../@journeyapps/wa-sqlite/dist/mc-wa-sqlite.js");
			return { default: factory };
		}, []);
		return factory();
	} else {
		const { default: factory } = await __vitePreload(async () => {
			const { default: factory } = await import("../../../../../../../@journeyapps/wa-sqlite/dist/wa-sqlite.js");
			return { default: factory };
		}, []);
		return factory();
	}
}
/**
* @internal
*/
var DEFAULT_MODULE_FACTORIES = {
	[WASQLiteVFS.IDBBatchAtomicVFS]: async (options) => {
		const module = await asyncModuleFactory(options.encryptionKey);
		const { IDBBatchAtomicVFS } = await __vitePreload(async () => {
			const { IDBBatchAtomicVFS } = await import("../../../../../../../@journeyapps/wa-sqlite/src/examples/IDBBatchAtomicVFS.js");
			return { IDBBatchAtomicVFS };
		}, __vite__mapDeps([0,1,2,3,4]));
		return {
			module,
			vfs: await IDBBatchAtomicVFS.create(options.dbFileName, module, { lockPolicy: "exclusive" })
		};
	},
	[WASQLiteVFS.AccessHandlePoolVFS]: async (options) => {
		const module = await syncModuleFactory(options.encryptionKey);
		const { AccessHandlePoolVFS } = await __vitePreload(async () => {
			const { AccessHandlePoolVFS } = await import("../../../../../../../@journeyapps/wa-sqlite/src/examples/AccessHandlePoolVFS.js");
			return { AccessHandlePoolVFS };
		}, __vite__mapDeps([5,1,2,3]));
		return {
			module,
			vfs: await AccessHandlePoolVFS.create(options.dbFileName, module)
		};
	},
	[WASQLiteVFS.OPFSCoopSyncVFS]: async (options) => {
		const module = await syncModuleFactory(options.encryptionKey);
		const { OPFSCoopSyncVFS } = await __vitePreload(async () => {
			const { OPFSCoopSyncVFS } = await import("../../../../../../../@journeyapps/wa-sqlite/src/examples/OPFSCoopSyncVFS.js");
			return { OPFSCoopSyncVFS };
		}, __vite__mapDeps([6,1,2,3]));
		return {
			module,
			vfs: await OPFSCoopSyncVFS.create(options.dbFileName, module)
		};
	},
	[WASQLiteVFS.OPFSWriteAheadVFS]: async (options) => {
		const module = await syncModuleFactory(options.encryptionKey);
		const { OPFSWriteAheadVFS } = await __vitePreload(async () => {
			const { OPFSWriteAheadVFS } = await import("../../../../../../../@journeyapps/wa-sqlite/src/examples/OPFSWriteAheadVFS.js");
			return { OPFSWriteAheadVFS };
		}, __vite__mapDeps([7,1,2,3,8,9,10]));
		return {
			module,
			vfs: await OPFSWriteAheadVFS.create(options.dbFileName, module, {})
		};
	}
};
//#endregion
export { DEFAULT_MODULE_FACTORIES, WASQLiteVFS, vfsRequiresDedicatedWorkers };
