// media-chrome/dist/menu/media-audio-track-menu-button.js
import { MediaUIAttributes } from "../constants.js";
import { t } from "../utils/i18n.js";
import { GlobalThis } from "../utils/server-safe-globals.js";
import { getMediaController, getStringAttr, setStringAttr } from "../utils/element-utils.js";
import { MediaChromeMenuButton } from "./media-chrome-menu-button.js";
//#region node_modules/media-chrome/dist/menu/media-audio-track-menu-button.js
var audioTrackIcon = `<svg aria-hidden="true" viewBox="0 0 24 24">
  <path d="M11 17H9.5V7H11v10Zm-3-3H6.5v-4H8v4Zm6-5h-1.5v6H14V9Zm3 7h-1.5V8H17v8Z"/>
  <path d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10Zm-2 0a8 8 0 1 0-16 0 8 8 0 0 0 16 0Z"/>
</svg>`;
function getSlotTemplateHTML() {
	return `
    <style>
      :host([aria-expanded="true"]) slot[name=tooltip] {
        display: none;
      }
    </style>
    <slot name="icon">${audioTrackIcon}</slot>
  `;
}
function getTooltipContentHTML() {
	return t("Audio");
}
var updateAriaLabel = (el) => {
	const label = t("Audio");
	el.setAttribute("aria-label", label);
};
var MediaAudioTrackMenuButton = class extends MediaChromeMenuButton {
	static get observedAttributes() {
		return [
			...super.observedAttributes,
			MediaUIAttributes.MEDIA_AUDIO_TRACK_ENABLED,
			MediaUIAttributes.MEDIA_AUDIO_TRACK_UNAVAILABLE
		];
	}
	connectedCallback() {
		super.connectedCallback();
		updateAriaLabel(this);
	}
	attributeChangedCallback(attrName, _oldValue, newValue) {
		super.attributeChangedCallback(attrName, _oldValue, newValue);
		if (attrName === MediaUIAttributes.MEDIA_LANG) updateAriaLabel(this);
	}
	/**
	* Returns the element with the id specified by the `invoketarget` attribute.
	* @return {HTMLElement | null}
	*/
	get invokeTargetElement() {
		var _a;
		if (this.invokeTarget != void 0) return super.invokeTargetElement;
		return (_a = getMediaController(this)) == null ? void 0 : _a.querySelector("media-audio-track-menu");
	}
	/**
	* Get enabled audio track id.
	* @return {string}
	*/
	get mediaAudioTrackEnabled() {
		var _a;
		return (_a = getStringAttr(this, MediaUIAttributes.MEDIA_AUDIO_TRACK_ENABLED)) != null ? _a : "";
	}
	set mediaAudioTrackEnabled(id) {
		setStringAttr(this, MediaUIAttributes.MEDIA_AUDIO_TRACK_ENABLED, id);
	}
};
MediaAudioTrackMenuButton.getSlotTemplateHTML = getSlotTemplateHTML;
MediaAudioTrackMenuButton.getTooltipContentHTML = getTooltipContentHTML;
if (!GlobalThis.customElements.get("media-audio-track-menu-button")) GlobalThis.customElements.define("media-audio-track-menu-button", MediaAudioTrackMenuButton);
//#endregion
export { MediaAudioTrackMenuButton };
