// media-chrome/dist/menu/media-playback-rate-menu-button.js
import { MediaUIAttributes } from "../constants.js";
import { t } from "../utils/i18n.js";
import { GlobalThis } from "../utils/server-safe-globals.js";
import { getMediaController, getNumericAttr, setNumericAttr } from "../utils/element-utils.js";
import { MediaChromeMenuButton } from "./media-chrome-menu-button.js";
function getSlotTemplateHTML(attrs) {
	return `
    <style>
      :host {
        min-width: 5ch;
        padding: var(--media-button-padding, var(--media-control-padding, 10px 5px));
      }
      
      :host([aria-expanded="true"]) slot {
        display: block;
      }

      :host([aria-expanded="true"]) slot[name=tooltip] {
        display: none;
      }
    </style>
    <slot name="icon">${attrs["mediaplaybackrate"] || 1}x</slot>
  `;
}
function getTooltipContentHTML() {
	return t("Playback rate");
}
var MediaPlaybackRateMenuButton = class extends MediaChromeMenuButton {
	static get observedAttributes() {
		return [...super.observedAttributes, MediaUIAttributes.MEDIA_PLAYBACK_RATE];
	}
	constructor() {
		var _a;
		super();
		this.container = this.shadowRoot.querySelector("slot[name=\"icon\"]");
		this.container.innerHTML = `${(_a = this.mediaPlaybackRate) != null ? _a : 1}x`;
	}
	attributeChangedCallback(attrName, oldValue, newValue) {
		super.attributeChangedCallback(attrName, oldValue, newValue);
		if (attrName === MediaUIAttributes.MEDIA_PLAYBACK_RATE) {
			const newPlaybackRate = newValue ? +newValue : NaN;
			const playbackRate = !Number.isNaN(newPlaybackRate) ? newPlaybackRate : 1;
			this.container.innerHTML = `${playbackRate}x`;
			this.setAttribute("aria-label", t("Playback rate {playbackRate}", { playbackRate }));
		}
	}
	/**
	* Returns the element with the id specified by the `invoketarget` attribute.
	*/
	get invokeTargetElement() {
		if (this.invokeTarget != void 0) return super.invokeTargetElement;
		return getMediaController(this).querySelector("media-playback-rate-menu");
	}
	/**
	* The current playback rate
	*/
	get mediaPlaybackRate() {
		return getNumericAttr(this, MediaUIAttributes.MEDIA_PLAYBACK_RATE, 1);
	}
	set mediaPlaybackRate(value) {
		setNumericAttr(this, MediaUIAttributes.MEDIA_PLAYBACK_RATE, value);
	}
};
MediaPlaybackRateMenuButton.getSlotTemplateHTML = getSlotTemplateHTML;
MediaPlaybackRateMenuButton.getTooltipContentHTML = getTooltipContentHTML;
if (!GlobalThis.customElements.get("media-playback-rate-menu-button")) GlobalThis.customElements.define("media-playback-rate-menu-button", MediaPlaybackRateMenuButton);
//#endregion
export { MediaPlaybackRateMenuButton };
