// @powersync/react/lib/hooks/watched/useSingleQuery.js
import React from "react";
//#region node_modules/@powersync/react/lib/hooks/watched/useSingleQuery.js
/**
* @internal not exported from `index.ts`
*/
var useSingleQuery = (options) => {
	const { query, powerSync, queryChanged, active } = options;
	const [output, setOutputState] = React.useState({
		isLoading: true,
		isFetching: true,
		data: [],
		error: void 0
	});
	const runQuery = React.useCallback(async (signal) => {
		setOutputState((prev) => ({
			...prev,
			isLoading: true,
			isFetching: true,
			error: void 0
		}));
		try {
			const compiledQuery = query.compile();
			const result = await query.execute({
				sql: compiledQuery.sql,
				parameters: [...compiledQuery.parameters],
				db: powerSync
			});
			if (signal.aborted) return;
			setOutputState((prev) => ({
				...prev,
				isLoading: false,
				isFetching: false,
				data: result,
				error: void 0
			}));
		} catch (error) {
			setOutputState((prev) => ({
				...prev,
				isLoading: false,
				isFetching: false,
				data: [],
				error
			}));
		}
	}, [queryChanged, query]);
	React.useEffect(() => {
		if (active) {
			const abortController = new AbortController();
			runQuery(abortController.signal);
			return () => {
				abortController.abort();
			};
		}
	}, [
		powerSync,
		active,
		queryChanged
	]);
	return {
		...output,
		refresh: runQuery
	};
};
//#endregion
export { useSingleQuery };
